// sanity/lib/_chunks/devAction-DaW6697o.js
'use strict';

var path = require('path');
var chalk = require('chalk');
var vite = require('vite');
var runtime = require('./runtime-CSusHM37.js');
var timing = require('./timing-DBaAPTKQ.js');
var servers = require('./servers-DtaNFRTM.js');
function _interopDefaultCompat(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    default: e
  };
}
var path__default = /*#__PURE__*/_interopDefaultCompat(path);
var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);
async function startDevServer(options) {
  const {
    cwd,
    httpPort,
    httpHost,
    basePath,
    reactStrictMode,
    vite: extendViteConfig
  } = options;
  const startTime = Date.now();
  runtime.debug("Writing Sanity runtime files");
  await runtime.writeSanityRuntime({
    cwd,
    reactStrictMode,
    watch: true,
    basePath
  });
  runtime.debug("Resolving vite config");
  const mode = "development";
  let viteConfig = await runtime.getViteConfig({
    basePath,
    mode: "development",
    server: {
      port: httpPort,
      host: httpHost
    },
    cwd
  });
  if (extendViteConfig) {
    viteConfig = await runtime.extendViteConfigWithUserConfig({
      command: "serve",
      mode
    }, viteConfig, extendViteConfig);
  }
  runtime.debug("Creating vite server");
  const server = await vite.createServer(viteConfig);
  const info = server.config.logger.info;
  runtime.debug("Listening on specified port");
  await server.listen();
  const startupDuration = Date.now() - startTime;
  const url = "http://".concat(httpHost || "localhost", ":").concat(httpPort || "3333").concat(basePath);
  info("Sanity Studio " + "using ".concat(chalk__default.default.cyan("vite@".concat(require("vite/package.json").version)), " ") + "ready in ".concat(chalk__default.default.cyan("".concat(Math.ceil(startupDuration), "ms")), " ") + "and running at ".concat(chalk__default.default.cyan(url)));
  return {
    close: () => server.close()
  };
}
async function startSanityDevServer(args, context) {
  const timers = timing.getTimer();
  const flags = args.extOptions;
  const {
    output,
    workDir,
    cliConfig
  } = context;
  timers.start("checkStudioDependencyVersions");
  timing.checkStudioDependencyVersions(workDir);
  timers.end("checkStudioDependencyVersions");
  if ((await timing.checkRequiredDependencies(context)).didInstall) {
    return;
  }
  const config = getDevServerConfig({
    flags,
    workDir,
    cliConfig,
    output
  });
  try {
    await startDevServer(config);
  } catch (err) {
    servers.gracefulServerDeath("dev", config.httpHost, config.httpPort, err);
  }
}
function getDevServerConfig(_ref) {
  let {
    flags,
    workDir,
    cliConfig,
    output
  } = _ref;
  var _a;
  const configSpinner = output.spinner("Checking configuration files...");
  const baseConfig = servers.getSharedServerConfig({
    flags,
    workDir,
    cliConfig
  });
  configSpinner.succeed();
  const env = process.env;
  const reactStrictMode = env.SANITY_STUDIO_REACT_STRICT_MODE ? env.SANITY_STUDIO_REACT_STRICT_MODE === "true" : Boolean(cliConfig == null ? void 0 : cliConfig.reactStrictMode);
  if (env.SANITY_STUDIO_BASEPATH && ((_a = cliConfig == null ? void 0 : cliConfig.project) == null ? void 0 : _a.basePath)) {
    output.warn("Overriding configured base path (".concat(cliConfig.project.basePath, ") with value from environment variable (").concat(env.SANITY_STUDIO_BASEPATH, ")"));
  }
  return {
    ...baseConfig,
    staticPath: path__default.default.join(workDir, "static"),
    reactStrictMode
  };
}
exports.default = startSanityDevServer;
