// sanity/lib/_chunks/devAction-F8mMTJzH.js
import path from 'path';
import chalk from 'chalk';
import { createServer } from 'vite';
import { debug, writeSanityRuntime, getViteConfig, extendViteConfigWithUserConfig } from './runtime-B_bGx0NP.js';
import { checkStudioDependencyVersions, checkRequiredDependencies, getTimer } from './timing-DjElMkC8.js';
import { gracefulServerDeath, getSharedServerConfig } from './servers-BEv-khCo.js';
async function startDevServer(options) {
  const {
    cwd,
    httpPort,
    httpHost,
    basePath,
    reactStrictMode,
    vite: extendViteConfig
  } = options;
  const startTime = Date.now();
  debug("Writing Sanity runtime files");
  await writeSanityRuntime({
    cwd,
    reactStrictMode,
    watch: true,
    basePath
  });
  debug("Resolving vite config");
  const mode = "development";
  let viteConfig = await getViteConfig({
    basePath,
    mode: "development",
    server: {
      port: httpPort,
      host: httpHost
    },
    cwd
  });
  if (extendViteConfig) {
    viteConfig = await extendViteConfigWithUserConfig({
      command: "serve",
      mode
    }, viteConfig, extendViteConfig);
  }
  debug("Creating vite server");
  const server = await createServer(viteConfig);
  const info = server.config.logger.info;
  debug("Listening on specified port");
  await server.listen();
  const startupDuration = Date.now() - startTime;
  const url = "http://".concat(httpHost || "localhost", ":").concat(httpPort || "3333").concat(basePath);
  info("Sanity Studio " + "using ".concat(chalk.cyan("vite@".concat(require("vite/package.json").version)), " ") + "ready in ".concat(chalk.cyan("".concat(Math.ceil(startupDuration), "ms")), " ") + "and running at ".concat(chalk.cyan(url)));
  return {
    close: () => server.close()
  };
}
async function startSanityDevServer(args, context) {
  const timers = getTimer();
  const flags = args.extOptions;
  const {
    output,
    workDir,
    cliConfig
  } = context;
  timers.start("checkStudioDependencyVersions");
  checkStudioDependencyVersions(workDir);
  timers.end("checkStudioDependencyVersions");
  if ((await checkRequiredDependencies(context)).didInstall) {
    return;
  }
  const config = getDevServerConfig({
    flags,
    workDir,
    cliConfig,
    output
  });
  try {
    await startDevServer(config);
  } catch (err) {
    gracefulServerDeath("dev", config.httpHost, config.httpPort, err);
  }
}
function getDevServerConfig(_ref) {
  let {
    flags,
    workDir,
    cliConfig,
    output
  } = _ref;
  var _a;
  const configSpinner = output.spinner("Checking configuration files...");
  const baseConfig = getSharedServerConfig({
    flags,
    workDir,
    cliConfig
  });
  configSpinner.succeed();
  const env = process.env;
  const reactStrictMode = env.SANITY_STUDIO_REACT_STRICT_MODE ? env.SANITY_STUDIO_REACT_STRICT_MODE === "true" : Boolean(cliConfig == null ? void 0 : cliConfig.reactStrictMode);
  if (env.SANITY_STUDIO_BASEPATH && ((_a = cliConfig == null ? void 0 : cliConfig.project) == null ? void 0 : _a.basePath)) {
    output.warn("Overriding configured base path (".concat(cliConfig.project.basePath, ") with value from environment variable (").concat(env.SANITY_STUDIO_BASEPATH, ")"));
  }
  return {
    ...baseConfig,
    staticPath: path.join(workDir, "static"),
    reactStrictMode
  };
}
export { startSanityDevServer as default };
