// sanity/lib/_chunks/index-CvUNfby3.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { useState, createElement, isValidElement } from 'react';
import { isValidElementType } from 'react-is';
import { useI18nText } from 'sanity';
import { Box } from '@sanity/ui';
import { Button } from './TooltipDelayGroupProvider-CvFQ5MgA.js';
import { usePane, useStructureTool, PaneHeader, PaneHeaderActions, BackLink, Pane } from './StructureToolProvider-DbWjiGhG.js';
import { ArrowLeftIcon } from '@sanity/icons';
import 'sanity/router';
import '@sanity/types';
import '@sanity/util/paths';
import 'lodash/omit.js';
import 'lodash/isNumber.js';
import 'lodash/isString.js';
import 'react-rx';
import styled from 'styled-components';
var __freeze = Object.freeze;
var __defProp = Object.defineProperty;
var __template = (cooked, raw) => __freeze(__defProp(cooked, "raw", {
  value: __freeze(raw || cooked.slice())
}));
var _a;
const Root = styled(Box)(_a || (_a = __template(["\n  position: relative;\n"])));
function UserComponentPaneContent(props) {
  const {
    children
  } = props;
  const {
    collapsed
  } = usePane();
  return /* @__PURE__ */jsx(Root, {
    hidden: collapsed,
    height: "fill",
    overflow: "auto",
    children
  });
}
function UserComponentPaneHeader(props) {
  const {
    actionHandlers,
    index,
    menuItems,
    menuItemGroups,
    title
  } = props;
  const {
    features
  } = useStructureTool();
  if (!(menuItems == null ? void 0 : menuItems.length) && !title) {
    return null;
  }
  return /* @__PURE__ */jsx(PaneHeader, {
    actions: /* @__PURE__ */jsx(PaneHeaderActions, {
      menuItems,
      menuItemGroups,
      actionHandlers
    }),
    backButton: features.backButton && index > 0 && /* @__PURE__ */jsx(Button, {
      as: BackLink,
      "data-as": "a",
      icon: ArrowLeftIcon,
      mode: "bleed",
      tooltipProps: {
        content: "Back"
      }
    }),
    title
  });
}
function UserComponentPane(props) {
  const {
    index,
    pane,
    paneKey,
    ...restProps
  } = props;
  const {
    child,
    component,
    menuItems,
    menuItemGroups,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    type: _unused,
    ...restPane
  } = pane;
  const [ref, setRef] = useState(null);
  const {
    title = ""
  } = useI18nText(pane);
  return /* @__PURE__ */jsxs(Pane, {
    id: paneKey,
    minWidth: 320,
    selected: restProps.isSelected,
    children: [/* @__PURE__ */jsx(UserComponentPaneHeader, {
      actionHandlers: ref == null ? void 0 : ref.actionHandlers,
      index,
      menuItems,
      menuItemGroups,
      title
    }), /* @__PURE__ */jsxs(UserComponentPaneContent, {
      children: [isValidElementType(component) && createElement(component, {
        ...restProps,
        ...restPane,
        // NOTE: here we're utilizing the function form of refs so setting
        // the ref causes a re-render for `UserComponentPaneHeader`
        ...{
          ref: setRef
        },
        child,
        // @todo: Fix typings
        paneKey
      }), isValidElement(component) && component]
    })]
  });
}
export { UserComponentPane as default };
