// sanity/lib/_chunks/index-D9-enJnZ.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var reactIs = require('react-is');
var sanity = require('sanity');
var ui = require('@sanity/ui');
var TooltipDelayGroupProvider = require('./TooltipDelayGroupProvider-CXQG_E3r.js');
var StructureToolProvider = require('./StructureToolProvider-DndCsu52.js');
var icons = require('@sanity/icons');
require('sanity/router');
require('@sanity/types');
require('@sanity/util/paths');
require('lodash/omit.js');
require('lodash/isNumber.js');
require('lodash/isString.js');
require('react-rx');
var styled = require('styled-components');
function _interopDefaultCompat(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    default: e
  };
}
var styled__default = /*#__PURE__*/_interopDefaultCompat(styled);
var __freeze = Object.freeze;
var __defProp = Object.defineProperty;
var __template = (cooked, raw) => __freeze(__defProp(cooked, "raw", {
  value: __freeze(raw || cooked.slice())
}));
var _a;
const Root = styled__default.default(ui.Box)(_a || (_a = __template(["\n  position: relative;\n"])));
function UserComponentPaneContent(props) {
  const {
    children
  } = props;
  const {
    collapsed
  } = StructureToolProvider.usePane();
  return /* @__PURE__ */jsxRuntime.jsx(Root, {
    hidden: collapsed,
    height: "fill",
    overflow: "auto",
    children
  });
}
function UserComponentPaneHeader(props) {
  const {
    actionHandlers,
    index,
    menuItems,
    menuItemGroups,
    title
  } = props;
  const {
    features
  } = StructureToolProvider.useStructureTool();
  if (!(menuItems == null ? void 0 : menuItems.length) && !title) {
    return null;
  }
  return /* @__PURE__ */jsxRuntime.jsx(StructureToolProvider.PaneHeader, {
    actions: /* @__PURE__ */jsxRuntime.jsx(StructureToolProvider.PaneHeaderActions, {
      menuItems,
      menuItemGroups,
      actionHandlers
    }),
    backButton: features.backButton && index > 0 && /* @__PURE__ */jsxRuntime.jsx(TooltipDelayGroupProvider.Button, {
      as: StructureToolProvider.BackLink,
      "data-as": "a",
      icon: icons.ArrowLeftIcon,
      mode: "bleed",
      tooltipProps: {
        content: "Back"
      }
    }),
    title
  });
}
function UserComponentPane(props) {
  const {
    index,
    pane,
    paneKey,
    ...restProps
  } = props;
  const {
    child,
    component,
    menuItems,
    menuItemGroups,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    type: _unused,
    ...restPane
  } = pane;
  const [ref, setRef] = React.useState(null);
  const {
    title = ""
  } = sanity.useI18nText(pane);
  return /* @__PURE__ */jsxRuntime.jsxs(StructureToolProvider.Pane, {
    id: paneKey,
    minWidth: 320,
    selected: restProps.isSelected,
    children: [/* @__PURE__ */jsxRuntime.jsx(UserComponentPaneHeader, {
      actionHandlers: ref == null ? void 0 : ref.actionHandlers,
      index,
      menuItems,
      menuItemGroups,
      title
    }), /* @__PURE__ */jsxRuntime.jsxs(UserComponentPaneContent, {
      children: [reactIs.isValidElementType(component) && React.createElement(component, {
        ...restProps,
        ...restPane,
        // NOTE: here we're utilizing the function form of refs so setting
        // the ref causes a re-render for `UserComponentPaneHeader`
        ...{
          ref: setRef
        },
        child,
        // @todo: Fix typings
        paneKey
      }), React.isValidElement(component) && component]
    })]
  });
}
exports.default = UserComponentPane;
