// sanity/lib/_chunks/index-Dkzo_-Bz.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { Box } from '@sanity/ui';
import { useGetI18nText, CommandList, useI18nText } from 'sanity';
import { useCallback } from 'react';
import { Button } from './TooltipDelayGroupProvider-CvFQ5MgA.js';
import { usePaneLayout, PaneItem, PaneContent, useStructureTool, usePane, PaneHeader, PaneHeaderActions, BackLink, Pane, _DEBUG } from './StructureToolProvider-DbWjiGhG.js';
import { ArrowLeftIcon } from '@sanity/icons';
import 'sanity/router';
import '@sanity/types';
import '@sanity/util/paths';
import 'lodash/omit.js';
import 'lodash/isNumber.js';
import 'lodash/isString.js';
import 'react-rx';
import styled from 'styled-components';
var __freeze = Object.freeze;
var __defProp = Object.defineProperty;
var __template = (cooked, raw) => __freeze(__defProp(cooked, "raw", {
  value: __freeze(raw || cooked.slice())
}));
var _a;
const Divider = styled.hr(_a || (_a = __template(["\n  background-color: var(--card-border-color);\n  height: 1px;\n  margin: 0;\n  border: none;\n"])));
function ListPaneContent(props) {
  const {
    childItemId,
    items,
    isActive,
    layout,
    showIcons,
    title
  } = props;
  const {
    collapsed: layoutCollapsed
  } = usePaneLayout();
  const getI18nText = useGetI18nText(items == null ? void 0 : items.filter(item => item.type !== "divider"));
  const getItemDisabled = useCallback(itemIndex => {
    var _a2;
    return ((_a2 = items == null ? void 0 : items.find((_, i) => i === itemIndex)) == null ? void 0 : _a2.type) === "divider";
  }, [items]);
  const shouldShowIconForItem = useCallback(item => {
    var _a2;
    const itemShowIcon = (_a2 = item.displayOptions) == null ? void 0 : _a2.showIcon;
    if (typeof itemShowIcon !== "undefined") {
      return itemShowIcon !== false;
    }
    return showIcons !== false;
  }, [showIcons]);
  const renderItem = useCallback((item, ctx) => {
    const {
      virtualIndex: itemIndex
    } = ctx;
    if (item.type === "divider") {
      return (
        // eslint-disable-next-line react/no-array-index-key
        /* @__PURE__ */
        jsx(Box, {
          marginBottom: 1,
          children: /* @__PURE__ */jsx(Divider, {})
        }, "divider-".concat(itemIndex))
      );
    }
    const pressed = !isActive && childItemId === item.id;
    const selected = isActive && childItemId === item.id;
    const value = item._id && item.schemaType ? {
      _id: item._id,
      _type: item.schemaType.name,
      title: item.title
    } : void 0;
    return /* @__PURE__ */jsx(PaneItem, {
      icon: shouldShowIconForItem(item) ? item.icon : false,
      id: item.id,
      layout,
      marginBottom: 1,
      pressed,
      schemaType: item.schemaType,
      selected,
      title: getI18nText(item).title,
      value
    }, item.id);
  }, [childItemId, getI18nText, isActive, layout, shouldShowIconForItem]);
  return /* @__PURE__ */jsx(PaneContent, {
    overflow: layoutCollapsed ? "hidden" : "auto",
    children: items && items.length > 0 && /* @__PURE__ */jsx(CommandList, {
      activeItemDataAttr: "data-hovered",
      ariaLabel: title,
      canReceiveFocus: true,
      getItemDisabled,
      itemHeight: 51,
      items,
      onlyShowSelectionWhenActive: true,
      paddingBottom: 1,
      paddingX: 3,
      renderItem,
      wrapAround: false
    })
  });
}
const ListPaneHeader = _ref => {
  let {
    index,
    menuItems,
    menuItemGroups,
    title
  } = _ref;
  const {
    features
  } = useStructureTool();
  const {
    collapsed,
    isLast
  } = usePane();
  const tabIndex = isLast && !collapsed ? -1 : 0;
  return /* @__PURE__ */jsx(PaneHeader, {
    actions: /* @__PURE__ */jsx(PaneHeaderActions, {
      menuItems,
      menuItemGroups
    }),
    backButton: features.backButton && index > 0 && /* @__PURE__ */jsx(Button, {
      as: BackLink,
      "data-as": "a",
      icon: ArrowLeftIcon,
      mode: "bleed",
      tooltipProps: {
        content: "Back"
      }
    }),
    tabIndex,
    title
  });
};
function ListPane(props) {
  const {
    childItemId,
    index,
    isActive,
    isSelected,
    pane,
    paneKey
  } = props;
  const {
    defaultLayout,
    displayOptions,
    items,
    menuItems,
    menuItemGroups
  } = pane;
  const showIcons = (displayOptions == null ? void 0 : displayOptions.showIcons) !== false;
  const {
    title
  } = useI18nText(pane);
  return /* @__PURE__ */jsxs(Pane, {
    currentMaxWidth: 350,
    "data-testid": "structure-tool-list-pane",
    "data-ui": "ListPane",
    id: paneKey,
    maxWidth: 640,
    minWidth: 320,
    selected: isSelected,
    children: [_DEBUG, /* @__PURE__ */jsx(ListPaneHeader, {
      index,
      menuItems,
      menuItemGroups,
      title
    }), /* @__PURE__ */jsx(ListPaneContent, {
      childItemId,
      isActive,
      items,
      layout: defaultLayout,
      showIcons,
      title
    }, paneKey)]
  });
}
export { ListPane as default };
