// sanity/lib/index.cjs.mjs
import cjs from './index.js';

export const ActiveWorkspaceMatcher = cjs.ActiveWorkspaceMatcher;
export const ActiveWorkspaceMatcherContext = cjs.ActiveWorkspaceMatcherContext;
export const AddonDatasetContext = cjs.AddonDatasetContext;
export const ArrayOfObjectOptionsInput = cjs.ArrayOfObjectOptionsInput;
export const ArrayOfObjectsFunctions = cjs.ArrayOfObjectsFunctions;
export const ArrayOfObjectsInput = cjs.ArrayOfObjectsInput;
export const ArrayOfObjectsItem = cjs.ArrayOfObjectsItem;
export const ArrayOfOptionsInput = cjs.ArrayOfOptionsInput;
export const ArrayOfPrimitiveOptionsInput = cjs.ArrayOfPrimitiveOptionsInput;
export const ArrayOfPrimitivesFunctions = cjs.ArrayOfPrimitivesFunctions;
export const ArrayOfPrimitivesInput = cjs.ArrayOfPrimitivesInput;
export const ArrayOfPrimitivesItem = cjs.ArrayOfPrimitivesItem;
export const AutoCollapseMenu = cjs.AutoCollapseMenu;
export const BlockEditor = cjs.BlockEditor;
export const BlockImagePreview = cjs.BlockImagePreview;
export const BlockPreview = cjs.BlockPreview;
export const BooleanInput = cjs.BooleanInput;
export const CONNECTING = cjs.CONNECTING;
export const ChangeBreadcrumb = cjs.ChangeBreadcrumb;
export const ChangeIndicator = cjs.ChangeIndicator;
export const ChangeIndicatorContext = cjs.ChangeIndicatorContext;
export const ChangeList = cjs.ChangeList;
export const ChangeResolver = cjs.ChangeResolver;
export const ChangeTitleSegment = cjs.ChangeTitleSegment;
export const CircularProgress = cjs.CircularProgress;
export const CollapseMenu = cjs.CollapseMenu;
export const CollapseMenuButton = cjs.CollapseMenuButton;
export const ColorSchemeCustomProvider = cjs.ColorSchemeCustomProvider;
export const ColorSchemeLocalStorageProvider = cjs.ColorSchemeLocalStorageProvider;
export const ColorSchemeProvider = cjs.ColorSchemeProvider;
export const ColorSchemeSetValueContext = cjs.ColorSchemeSetValueContext;
export const ColorSchemeValueContext = cjs.ColorSchemeValueContext;
export const CommandList = cjs.CommandList;
export const CompactPreview = cjs.CompactPreview;
export const ConfigPropertyError = cjs.ConfigPropertyError;
export const ConfigResolutionError = cjs.ConfigResolutionError;
export const ConnectorContext = cjs.ConnectorContext;
export const ContextMenuButton = cjs.ContextMenuButton;
export const CorsOriginError = cjs.CorsOriginError;
export const CrossDatasetReferenceInput = cjs.CrossDatasetReferenceInput;
export const DEFAULT_MAX_RECURSION_DEPTH = cjs.DEFAULT_MAX_RECURSION_DEPTH;
export const DEFAULT_STUDIO_CLIENT_OPTIONS = cjs.DEFAULT_STUDIO_CLIENT_OPTIONS;
export const DRAFTS_FOLDER = cjs.DRAFTS_FOLDER;
export const DateInput = cjs.DateInput;
export const DateTimeInput = cjs.DateTimeInput;
export const DefaultPreview = cjs.DefaultPreview;
export const DetailPreview = cjs.DetailPreview;
export const DiffCard = cjs.DiffCard;
export const DiffContext = cjs.DiffContext;
export const DiffErrorBoundary = cjs.DiffErrorBoundary;
export const DiffFromTo = cjs.DiffFromTo;
export const DiffInspectWrapper = cjs.DiffInspectWrapper;
export const DiffString = cjs.DiffString;
export const DiffStringSegment = cjs.DiffStringSegment;
export const DiffTooltip = cjs.DiffTooltip;
export const DocumentChangeContext = cjs.DocumentChangeContext;
export const DocumentPreviewPresence = cjs.DocumentPreviewPresence;
export const DocumentStatus = cjs.DocumentStatus;
export const DocumentStatusIndicator = cjs.DocumentStatusIndicator;
export const EMPTY_ARRAY = cjs.EMPTY_ARRAY;
export const EMPTY_OBJECT = cjs.EMPTY_OBJECT;
export const EmailInput = cjs.EmailInput;
export const FallbackDiff = cjs.FallbackDiff;
export const FieldActionMenu = cjs.FieldActionMenu;
export const FieldActionsProvider = cjs.FieldActionsProvider;
export const FieldActionsResolver = cjs.FieldActionsResolver;
export const FieldChange = cjs.FieldChange;
export const FieldPresence = cjs.FieldPresence;
export const FieldPresenceInner = cjs.FieldPresenceInner;
export const FieldPresenceWithOverlay = cjs.FieldPresenceWithOverlay;
export const FieldPresenceWithoutOverlay = cjs.FieldPresenceWithoutOverlay;
export const FileInput = cjs.FileInput;
export const FormBuilderContext = cjs.FormBuilderContext;
export const FormCallbacksContext = cjs.FormCallbacksContext;
export const FormCallbacksProvider = cjs.FormCallbacksProvider;
export const FormField = cjs.FormField;
export const FormFieldHeaderText = cjs.FormFieldHeaderText;
export const FormFieldPresenceContext = cjs.FormFieldPresenceContext;
export const FormFieldSet = cjs.FormFieldSet;
export const FormFieldStatus = cjs.FormFieldStatus;
export const FormFieldValidationStatus = cjs.FormFieldValidationStatus;
export const FormInput = cjs.FormInput;
export const FormValueProvider = cjs.FormValueProvider;
export const FromTo = cjs.FromTo;
export const FromToArrow = cjs.FromToArrow;
export const GroupChange = cjs.GroupChange;
export const ImageInput = cjs.ImageInput;
export const ImperativeToast = cjs.ImperativeToast;
export const InlinePreview = cjs.InlinePreview;
export const InsufficientPermissionsMessage = cjs.InsufficientPermissionsMessage;
export const LegacyLayerProvider = cjs.LegacyLayerProvider;
export const LinearProgress = cjs.LinearProgress;
export const LoadingBlock = cjs.LoadingBlock;
export const LocaleProvider = cjs.LocaleProvider;
export const LocaleProviderBase = cjs.LocaleProviderBase;
export const MediaPreview = cjs.MediaPreview;
export const MemberField = cjs.MemberField;
export const MemberFieldError = cjs.MemberFieldError;
export const MemberFieldSet = cjs.MemberFieldSet;
export const MemberItemError = cjs.MemberItemError;
export const MetaInfo = cjs.MetaInfo;
export const NavbarContext = cjs.NavbarContext;
export const NoChanges = cjs.NoChanges;
export const NumberInput = cjs.NumberInput;
export const ObjectInput = cjs.ObjectInput;
export const ObjectInputMember = cjs.ObjectInputMember;
export const ObjectInputMembers = cjs.ObjectInputMembers;
export const ObjectMembers = cjs.ObjectMembers;
export const OverlayDisabled = cjs.OverlayDisabled;
export const PatchEvent = cjs.PatchEvent;
export const PopoverDialog = cjs.PopoverDialog;
export const PortableTextInput = cjs.PortableTextInput;
export const PresenceOverlay = cjs.PresenceOverlay;
export const Preview = cjs.Preview;
export const PreviewCard = cjs.PreviewCard;
export const PreviewLoader = cjs.PreviewLoader;
export const ReferenceInput = cjs.ReferenceInput;
export const ReferenceInputOptionsProvider = cjs.ReferenceInputOptionsProvider;
export const ResourceCacheProvider = cjs.ResourceCacheProvider;
export const RevertChangesButton = cjs.RevertChangesButton;
export const ReviewChangesContext = cjs.ReviewChangesContext;
export const SANITY_PATCH_TYPE = cjs.SANITY_PATCH_TYPE;
export const SANITY_VERSION = cjs.SANITY_VERSION;
export const SESSION_ID = cjs.SESSION_ID;
export const SanityDefaultPreview = cjs.SanityDefaultPreview;
export const SchemaError = cjs.SchemaError;
export const ScrollContainer = cjs.ScrollContainer;
export const SearchButton = cjs.SearchButton;
export const SearchDialog = cjs.SearchDialog;
export const SelectInput = cjs.SelectInput;
export const SlugInput = cjs.SlugInput;
export const SourceProvider = cjs.SourceProvider;
export const StatusButton = cjs.StatusButton;
export const StringInput = cjs.StringInput;
export const StudioLayout = cjs.StudioLayout;
export const StudioLayoutComponent = cjs.StudioLayoutComponent;
export const StudioNavbar = cjs.StudioNavbar;
export const StudioProvider = cjs.StudioProvider;
export const StudioToolMenu = cjs.StudioToolMenu;
export const TagsArrayInput = cjs.TagsArrayInput;
export const TemplatePreview = cjs.TemplatePreview;
export const TextInput = cjs.TextInput;
export const TextWithTone = cjs.TextWithTone;
export const Timeline = cjs.Timeline;
export const TimelineController = cjs.TimelineController;
export const ToolLink = cjs.ToolLink;
export const TooltipOfDisabled = cjs.TooltipOfDisabled;
export const Tracker = cjs.Tracker;
export const TransformPatches = cjs.TransformPatches;
export const Translate = cjs.Translate;
export const UniversalArrayInput = cjs.UniversalArrayInput;
export const UpsellDescriptionSerializer = cjs.UpsellDescriptionSerializer;
export const UpsellDialogDismissed = cjs.UpsellDialogDismissed;
export const UpsellDialogLearnMoreCtaClicked = cjs.UpsellDialogLearnMoreCtaClicked;
export const UpsellDialogUpgradeCtaClicked = cjs.UpsellDialogUpgradeCtaClicked;
export const UpsellDialogViewed = cjs.UpsellDialogViewed;
export const UrlInput = cjs.UrlInput;
export const UserAvatar = cjs.UserAvatar;
export const UserColorManagerProvider = cjs.UserColorManagerProvider;
export const ValueError = cjs.ValueError;
export const VirtualizerScrollInstanceContext = cjs.VirtualizerScrollInstanceContext;
export const VirtualizerScrollInstanceProvider = cjs.VirtualizerScrollInstanceProvider;
export const WorkspaceProvider = cjs.WorkspaceProvider;
export const WorkspacesContext = cjs.WorkspacesContext;
export const WorkspacesProvider = cjs.WorkspacesProvider;
export const __tmp_wrap_presenceStore = cjs.__tmp_wrap_presenceStore;
export const _createAuthStore = cjs._createAuthStore;
export const _useColorSchemeInternalValue = cjs._useColorSchemeInternalValue;
export const asLoadable = cjs.asLoadable;
export const checkoutPair = cjs.checkoutPair;
export const collate = cjs.collate;
export const createAuthStore = cjs.createAuthStore;
export const createBufferedDocument = cjs.createBufferedDocument;
export const createConnectionStatusStore = cjs.createConnectionStatusStore;
export const createDocumentPreviewStore = cjs.createDocumentPreviewStore;
export const createDocumentStore = cjs.createDocumentStore;
export const createDraftFrom = cjs.createDraftFrom;
export const createGrantsStore = cjs.createGrantsStore;
export const createHistoryStore = cjs.createHistoryStore;
export const createHookFromObservableFactory = cjs.createHookFromObservableFactory;
export const createKeyValueStore = cjs.createKeyValueStore;
export const createObservableBufferedDocument = cjs.createObservableBufferedDocument;
export const createProjectStore = cjs.createProjectStore;
export const createPublishedFrom = cjs.createPublishedFrom;
export const createSchema = cjs.createSchema;
export const createSearchQuery = cjs.createSearchQuery;
export const createSharedResizeObserver = cjs.createSharedResizeObserver;
export const createTrackerScope = cjs.createTrackerScope;
export const createUseReporter = cjs.createUseReporter;
export const createUserColorManager = cjs.createUserColorManager;
export const createUserStore = cjs.createUserStore;
export const dec = cjs.dec;
export const defaultLocale = cjs.defaultLocale;
export const defaultTheme = cjs.defaultTheme;
export const defineLocale = cjs.defineLocale;
export const defineLocaleResourceBundle = cjs.defineLocaleResourceBundle;
export const defineLocalesResources = cjs.defineLocalesResources;
export const defineSearchFilter = cjs.defineSearchFilter;
export const defineSearchFilterOperators = cjs.defineSearchFilterOperators;
export const defineSearchOperator = cjs.defineSearchOperator;
export const diffMatchPatch = cjs.diffMatchPatch;
export const documentFieldActionsReducer = cjs.documentFieldActionsReducer;
export const documentIdEquals = cjs.documentIdEquals;
export const editState = cjs.editState;
export const emitOperation = cjs.emitOperation;
export const escapeField = cjs.escapeField;
export const fieldNeedsEscape = cjs.fieldNeedsEscape;
export const findIndex = cjs.findIndex;
export const flattenConfig = cjs.flattenConfig;
export const getAnnotationAtPath = cjs.getAnnotationAtPath;
export const getAnnotationColor = cjs.getAnnotationColor;
export const getDiffAtPath = cjs.getDiffAtPath;
export const getDocumentPairPermissions = cjs.getDocumentPairPermissions;
export const getDocumentValuePermissions = cjs.getDocumentValuePermissions;
export const getDraftId = cjs.getDraftId;
export const getIdPair = cjs.getIdPair;
export const getInitialValueStream = cjs.getInitialValueStream;
export const getItemKey = cjs.getItemKey;
export const getItemKeySegment = cjs.getItemKeySegment;
export const getNamelessWorkspaceIdentifier = cjs.getNamelessWorkspaceIdentifier;
export const getPairListener = cjs.getPairListener;
export const getPreviewPaths = cjs.getPreviewPaths;
export const getPreviewStateObservable = cjs.getPreviewStateObservable;
export const getPreviewValueWithFallback = cjs.getPreviewValueWithFallback;
export const getProviderTitle = cjs.getProviderTitle;
export const getPublishedId = cjs.getPublishedId;
export const getSchemaTypeTitle = cjs.getSchemaTypeTitle;
export const getTemplatePermissions = cjs.getTemplatePermissions;
export const getValueAtPath = cjs.getValueAtPath;
export const getValueError = cjs.getValueError;
export const getWorkspaceIdentifier = cjs.getWorkspaceIdentifier;
export const globalScope = cjs.globalScope;
export const grantsPermissionOn = cjs.grantsPermissionOn;
export const inc = cjs.inc;
export const initialDocumentFieldActions = cjs.initialDocumentFieldActions;
export const insert = cjs.insert;
export const isAddedItemDiff = cjs.isAddedItemDiff;
export const isArrayOfBlocksInputProps = cjs.isArrayOfBlocksInputProps;
export const isArrayOfObjectsInputProps = cjs.isArrayOfObjectsInputProps;
export const isArrayOfPrimitivesInputProps = cjs.isArrayOfPrimitivesInputProps;
export const isAuthStore = cjs.isAuthStore;
export const isBooleanInputProps = cjs.isBooleanInputProps;
export const isBuilder = cjs.isBuilder;
export const isCookielessCompatibleLoginMethod = cjs.isCookielessCompatibleLoginMethod;
export const isDev = cjs.isDev;
export const isDraft = cjs.isDraft;
export const isDraftId = cjs.isDraftId;
export const isEmptyObject = cjs.isEmptyObject;
export const isFieldChange = cjs.isFieldChange;
export const isGroupChange = cjs.isGroupChange;
export const isNonNullable = cjs.isNonNullable;
export const isNumberInputProps = cjs.isNumberInputProps;
export const isObjectInputProps = cjs.isObjectInputProps;
export const isObjectItemProps = cjs.isObjectItemProps;
export const isProd = cjs.isProd;
export const isPublishedId = cjs.isPublishedId;
export const isRecord = cjs.isRecord;
export const isRemovedItemDiff = cjs.isRemovedItemDiff;
export const isString = cjs.isString;
export const isStringInputProps = cjs.isStringInputProps;
export const isUnchangedDiff = cjs.isUnchangedDiff;
export const joinPath = cjs.joinPath;
export const listenQuery = cjs.listenQuery;
export const matchWorkspace = cjs.matchWorkspace;
export const newDraftFrom = cjs.newDraftFrom;
export const noop = cjs.noop;
export const normalizeIndexSegment = cjs.normalizeIndexSegment;
export const normalizeIndexTupleSegment = cjs.normalizeIndexTupleSegment;
export const normalizeKeySegment = cjs.normalizeKeySegment;
export const normalizePathSegment = cjs.normalizePathSegment;
export const onRetry = cjs.onRetry;
export const operationEvents = cjs.operationEvents;
export const operatorDefinitions = cjs.operatorDefinitions;
export const pathToString = cjs.pathToString;
export const pathsAreEqual = cjs.pathsAreEqual;
export const prefixPath = cjs.prefixPath;
export const prepareConfig = cjs.prepareConfig;
export const prepareForPreview = cjs.prepareForPreview;
export const removeDupes = cjs.removeDupes;
export const removeMissingReferences = cjs.removeMissingReferences;
export const removeUndefinedLocaleResources = cjs.removeUndefinedLocaleResources;
export const resizeObserver = cjs.resizeObserver;
export const resolveDiffComponent = cjs.resolveDiffComponent;
export const resolveInitialObjectValue = cjs.resolveInitialObjectValue;
export const resolveInitialValue = cjs.resolveInitialValue;
export const resolveInitialValueForType = cjs.resolveInitialValueForType;
export const resolveSchemaTypes = cjs.resolveSchemaTypes;
export const set = cjs.set;
export const setIfMissing = cjs.setIfMissing;
export const sliceString = cjs.sliceString;
export const snapshotPair = cjs.snapshotPair;
export const stringToPath = cjs.stringToPath;
export const supportsTouch = cjs.supportsTouch;
export const truncateString = cjs.truncateString;
export const unset = cjs.unset;
export const unstable_useValuePreview = cjs.unstable_useValuePreview;
export const usEnglishLocale = cjs.usEnglishLocale;
export const useActiveWorkspace = cjs.useActiveWorkspace;
export const useAnnotationColor = cjs.useAnnotationColor;
export const useClient = cjs.useClient;
export const useColorScheme = cjs.useColorScheme;
export const useColorSchemeOptions = cjs.useColorSchemeOptions;
export const useColorSchemeSetValue = cjs.useColorSchemeSetValue;
export const useColorSchemeValue = cjs.useColorSchemeValue;
export const useConnectionStatusStore = cjs.useConnectionStatusStore;
export const useCurrentLocale = cjs.useCurrentLocale;
export const useCurrentUser = cjs.useCurrentUser;
export const useDataset = cjs.useDataset;
export const useDateTimeFormat = cjs.useDateTimeFormat;
export const useDidUpdate = cjs.useDidUpdate;
export const useDiffAnnotationColor = cjs.useDiffAnnotationColor;
export const useDocumentChange = cjs.useDocumentChange;
export const useDocumentOperation = cjs.useDocumentOperation;
export const useDocumentPairPermissions = cjs.useDocumentPairPermissions;
export const useDocumentPairPermissionsFromHookFactory = cjs.useDocumentPairPermissionsFromHookFactory;
export const useDocumentPresence = cjs.useDocumentPresence;
export const useDocumentPreviewStore = cjs.useDocumentPreviewStore;
export const useDocumentStore = cjs.useDocumentStore;
export const useDocumentValuePermissions = cjs.useDocumentValuePermissions;
export const useDocumentValuePermissionsFromHookFactory = cjs.useDocumentValuePermissionsFromHookFactory;
export const useDocumentValues = cjs.useDocumentValues;
export const useFeatureEnabled = cjs.useFeatureEnabled;
export const useFieldActions = cjs.useFieldActions;
export const useFormBuilder = cjs.useFormBuilder;
export const useFormCallbacks = cjs.useFormCallbacks;
export const useFormValue = cjs.useFormValue;
export const useGetI18nText = cjs.useGetI18nText;
export const useGlobalPresence = cjs.useGlobalPresence;
export const useGrantsStore = cjs.useGrantsStore;
export const useHistoryStore = cjs.useHistoryStore;
export const useHoveredField = cjs.useHoveredField;
export const useI18nText = cjs.useI18nText;
export const useInitialValue = cjs.useInitialValue;
export const useInitialValueResolverContext = cjs.useInitialValueResolverContext;
export const useKeyValueStore = cjs.useKeyValueStore;
export const useListFormat = cjs.useListFormat;
export const useLoadable = cjs.useLoadable;
export const useLocale = cjs.useLocale;
export const useMiddlewareComponents = cjs.useMiddlewareComponents;
export const usePresenceStore = cjs.usePresenceStore;
export const usePreviewCard = cjs.usePreviewCard;
export const useProjectId = cjs.useProjectId;
export const useProjectStore = cjs.useProjectStore;
export const useReferenceInputOptions = cjs.useReferenceInputOptions;
export const useRelativeTime = cjs.useRelativeTime;
export const useReportedValues = cjs.useReportedValues;
export const useReporter = cjs.useReporter;
export const useResolveInitialValueForType = cjs.useResolveInitialValueForType;
export const useResourceCache = cjs.useResourceCache;
export const useRovingFocus = cjs.useRovingFocus;
export const useSchema = cjs.useSchema;
export const useSource = cjs.useSource;
export const useTemplatePermissions = cjs.useTemplatePermissions;
export const useTemplatePermissionsFromHookFactory = cjs.useTemplatePermissionsFromHookFactory;
export const useTemplates = cjs.useTemplates;
export const useTranslation = cjs.useTranslation;
export const useUnique = cjs.useUnique;
export const useUnitFormatter = cjs.useUnitFormatter;
export const useUser = cjs.useUser;
export const useUserColor = cjs.useUserColor;
export const useUserColorManager = cjs.useUserColorManager;
export const useUserStore = cjs.useUserStore;
export const useVirtualizerScrollInstance = cjs.useVirtualizerScrollInstance;
export const useWorkspace = cjs.useWorkspace;
export const useWorkspaces = cjs.useWorkspaces;
export const useZIndex = cjs.useZIndex;
export const userHasRole = cjs.userHasRole;
export const validateBasePaths = cjs.validateBasePaths;
export const validateDocument = cjs.validateDocument;
export const validateNames = cjs.validateNames;
export const validateWorkspaces = cjs.validateWorkspaces;
export const validation = cjs.validation;
export const visitDiff = cjs.visitDiff;
export const AddonDatasetProvider = cjs.AddonDatasetProvider;
export const ArrayOfObjectsInputMember = cjs.ArrayOfObjectsInputMember;
export const ArrayOfObjectsInputMembers = cjs.ArrayOfObjectsInputMembers;
export const BetaBadge = cjs.BetaBadge;
export const ChangeConnectorRoot = cjs.ChangeConnectorRoot;
export const ChangeFieldWrapper = cjs.ChangeFieldWrapper;
export const DefaultDocument = cjs.DefaultDocument;
export const DisabledChangeConnectorRoot = cjs.DisabledChangeConnectorRoot;
export const EnabledChangeConnectorRoot = cjs.EnabledChangeConnectorRoot;
export const FormBuilder = cjs.FormBuilder;
export const FormProvider = cjs.FormProvider;
export const GetHookCollectionState = cjs.GetHookCollectionState;
export const GlobalErrorHandler = cjs.GlobalErrorHandler;
export const Hotkeys = cjs.Hotkeys;
export const HoveredFieldProvider = cjs.HoveredFieldProvider;
export const IntentButton = cjs.IntentButton;
export const PresenceScope = cjs.PresenceScope;
export const RelativeTime = cjs.RelativeTime;
export const ScrollMonitor = cjs.ScrollMonitor;
export const Studio = cjs.Studio;
export const StudioLogo = cjs.StudioLogo;
export const TelephoneInput = cjs.TelephoneInput;
export const WithReferringDocuments = cjs.WithReferringDocuments;
export const ZIndexProvider = cjs.ZIndexProvider;
export const _isCustomDocumentTypeDefinition = cjs._isCustomDocumentTypeDefinition;
export const _isSanityDocumentTypeDefinition = cjs._isSanityDocumentTypeDefinition;
export const buildLegacyTheme = cjs.buildLegacyTheme;
export const createConfig = cjs.createConfig;
export const createMockAuthStore = cjs.createMockAuthStore;
export const createNoopTrackerScope = cjs.createNoopTrackerScope;
export const createPatchChannel = cjs.createPatchChannel;
export const createPlugin = cjs.createPlugin;
export const createSourceFromConfig = cjs.createSourceFromConfig;
export const createWorkspaceFromConfig = cjs.createWorkspaceFromConfig;
export const decodePath = cjs.decodePath;
export const defaultRenderAnnotation = cjs.defaultRenderAnnotation;
export const defaultRenderBlock = cjs.defaultRenderBlock;
export const defaultRenderField = cjs.defaultRenderField;
export const defaultRenderInlineBlock = cjs.defaultRenderInlineBlock;
export const defaultRenderInput = cjs.defaultRenderInput;
export const defaultRenderItem = cjs.defaultRenderItem;
export const defaultRenderPreview = cjs.defaultRenderPreview;
export const defaultTemplateForType = cjs.defaultTemplateForType;
export const defaultTemplatesForSchema = cjs.defaultTemplatesForSchema;
export const defineConfig = cjs.defineConfig;
export const defineDocumentFieldAction = cjs.defineDocumentFieldAction;
export const defineDocumentInspector = cjs.defineDocumentInspector;
export const definePlugin = cjs.definePlugin;
export const diffResolver = cjs.diffResolver;
export const encodePath = cjs.encodePath;
export const fromMutationPatches = cjs.fromMutationPatches;
export const getConfigContextFromSource = cjs.getConfigContextFromSource;
export const getExpandOperations = cjs.getExpandOperations;
export const isArray = cjs.isArray;
export const isTruthy = cjs.isTruthy;
export const prepareTemplates = cjs.prepareTemplates;
export const remoteSnapshots = cjs.remoteSnapshots;
export const renderStudio = cjs.renderStudio;
export const resolveConditionalProperty = cjs.resolveConditionalProperty;
export const resolveConfig = cjs.resolveConfig;
export const setAtPath = cjs.setAtPath;
export const toMutationPatches = cjs.toMutationPatches;
export const uncaughtErrorHandler = cjs.uncaughtErrorHandler;
export const useAddonDataset = cjs.useAddonDataset;
export const useConfigContextFromSource = cjs.useConfigContextFromSource;
export const useConnectionState = cjs.useConnectionState;
export const useDocumentOperationEvent = cjs.useDocumentOperationEvent;
export const useDocumentType = cjs.useDocumentType;
export const useEditState = cjs.useEditState;
export const useFormState = cjs.useFormState;
export const useFormattedDuration = cjs.useFormattedDuration;
export const useNumberFormat = cjs.useNumberFormat;
export const useOnScroll = cjs.useOnScroll;
export const useProject = cjs.useProject;
export const useProjectDatasets = cjs.useProjectDatasets;
export const useReviewChanges = cjs.useReviewChanges;
export const useSyncState = cjs.useSyncState;
export const useThrottledCallback = cjs.useThrottledCallback;
export const useTimeAgo = cjs.useTimeAgo;
export const useTimelineSelector = cjs.useTimelineSelector;
export const useTimelineStore = cjs.useTimelineStore;
export const useTools = cjs.useTools;
export const useUserListWithPermissions = cjs.useUserListWithPermissions;
export const useValidationStatus = cjs.useValidationStatus;
export const defineArrayMember = cjs.defineArrayMember;
export const defineField = cjs.defineField;
export const defineType = cjs.defineType;
export const isArrayOfBlocksSchemaType = cjs.isArrayOfBlocksSchemaType;
export const isArrayOfObjectsSchemaType = cjs.isArrayOfObjectsSchemaType;
export const isArrayOfPrimitivesSchemaType = cjs.isArrayOfPrimitivesSchemaType;
export const isArraySchemaType = cjs.isArraySchemaType;
export const isBlockChildrenObjectField = cjs.isBlockChildrenObjectField;
export const isBlockListObjectField = cjs.isBlockListObjectField;
export const isBlockSchemaType = cjs.isBlockSchemaType;
export const isBlockStyleObjectField = cjs.isBlockStyleObjectField;
export const isBooleanSchemaType = cjs.isBooleanSchemaType;
export const isCreateIfNotExistsMutation = cjs.isCreateIfNotExistsMutation;
export const isCreateMutation = cjs.isCreateMutation;
export const isCreateOrReplaceMutation = cjs.isCreateOrReplaceMutation;
export const isCreateSquashedMutation = cjs.isCreateSquashedMutation;
export const isCrossDatasetReference = cjs.isCrossDatasetReference;
export const isCrossDatasetReferenceSchemaType = cjs.isCrossDatasetReferenceSchemaType;
export const isDeleteMutation = cjs.isDeleteMutation;
export const isDeprecatedSchemaType = cjs.isDeprecatedSchemaType;
export const isDeprecationConfiguration = cjs.isDeprecationConfiguration;
export const isDocumentSchemaType = cjs.isDocumentSchemaType;
export const isImage = cjs.isImage;
export const isIndexSegment = cjs.isIndexSegment;
export const isIndexTuple = cjs.isIndexTuple;
export const isKeySegment = cjs.isKeySegment;
export const isKeyedObject = cjs.isKeyedObject;
export const isNumberSchemaType = cjs.isNumberSchemaType;
export const isObjectSchemaType = cjs.isObjectSchemaType;
export const isPatchMutation = cjs.isPatchMutation;
export const isPortableTextListBlock = cjs.isPortableTextListBlock;
export const isPortableTextSpan = cjs.isPortableTextSpan;
export const isPortableTextTextBlock = cjs.isPortableTextTextBlock;
export const isPrimitiveSchemaType = cjs.isPrimitiveSchemaType;
export const isReference = cjs.isReference;
export const isReferenceSchemaType = cjs.isReferenceSchemaType;
export const isSanityDocument = cjs.isSanityDocument;
export const isSlug = cjs.isSlug;
export const isSpanSchemaType = cjs.isSpanSchemaType;
export const isStringSchemaType = cjs.isStringSchemaType;
export const isTitledListValue = cjs.isTitledListValue;
export const isTypedObject = cjs.isTypedObject;
export const isValidationError = cjs.isValidationError;
export const isValidationErrorMarker = cjs.isValidationErrorMarker;
export const isValidationInfo = cjs.isValidationInfo;
export const isValidationInfoMarker = cjs.isValidationInfoMarker;
export const isValidationWarning = cjs.isValidationWarning;
export const isValidationWarningMarker = cjs.isValidationWarningMarker;
export const typed = cjs.typed;
