// sanity/lib/structure.esm.js
export { CommentsIntentProvider, ComponentBuilder, ComponentViewBuilder, ConfirmDeleteDialogContainer as ConfirmDeleteDialog, DEFAULT_INTENT_HANDLER, DocumentBuilder, DocumentInspectorHeader, DocumentListBuilder, DocumentListItemBuilder, DocumentListPane, DocumentPane, DocumentPaneProvider, DocumentTypeListBuilder, FormViewBuilder, GenericListBuilder, GenericViewBuilder, HELP_URL, InitialValueTemplateItemBuilder, ListBuilder, ListItemBuilder, MenuItemBuilder, MenuItemGroupBuilder, PaneLayout, PaneRouterContext, SerializeError, StructureToolProvider, component, createStructureBuilder, defaultInitialValueTemplateItems, defaultIntentChecker, documentFromEditor, documentFromEditorWithInitialValue, form, getOrderingMenuItem, getOrderingMenuItemsForSchemaType, getTypeNamesFromFilter, isDocumentListItem, maybeSerializeInitialValueTemplateItem, maybeSerializeMenuItem, maybeSerializeMenuItemGroup, maybeSerializeView, menuItemsFromInitialValueTemplateItems, shallowIntentChecker, structureLocaleNamespace, structureTool, useDocumentPane, useDocumentTitle, usePaneRouter, useStructureTool } from './_chunks/StructureToolProvider-DbWjiGhG.js';
import 'react/jsx-runtime';
import '@sanity/icons';
import '@sanity/ui';
import 'react';
import './_chunks/TooltipDelayGroupProvider-CvFQ5MgA.js';
import 'sanity';
import '@sanity/telemetry/react';
import 'lodash/template.js';
import 'lodash/isEqual.js';
import 'sanity/router';
import 'react-rx';
import 'rxjs';
import '@sanity/types';
import '@sanity/util/paths';
import 'lodash/omit.js';
import 'lodash/isNumber.js';
import 'lodash/isString.js';
import '@sanity/uuid';
import 'lodash/findIndex.js';
