// @sanity/types/lib/index.cjs.mjs
import cjs from './index.js';

export const defineArrayMember = cjs.defineArrayMember;
export const defineField = cjs.defineField;
export const defineType = cjs.defineType;
export const isArrayOfBlocksSchemaType = cjs.isArrayOfBlocksSchemaType;
export const isArrayOfObjectsSchemaType = cjs.isArrayOfObjectsSchemaType;
export const isArrayOfPrimitivesSchemaType = cjs.isArrayOfPrimitivesSchemaType;
export const isArraySchemaType = cjs.isArraySchemaType;
export const isBlockChildrenObjectField = cjs.isBlockChildrenObjectField;
export const isBlockListObjectField = cjs.isBlockListObjectField;
export const isBlockSchemaType = cjs.isBlockSchemaType;
export const isBlockStyleObjectField = cjs.isBlockStyleObjectField;
export const isBooleanSchemaType = cjs.isBooleanSchemaType;
export const isCreateIfNotExistsMutation = cjs.isCreateIfNotExistsMutation;
export const isCreateMutation = cjs.isCreateMutation;
export const isCreateOrReplaceMutation = cjs.isCreateOrReplaceMutation;
export const isCreateSquashedMutation = cjs.isCreateSquashedMutation;
export const isCrossDatasetReference = cjs.isCrossDatasetReference;
export const isCrossDatasetReferenceSchemaType = cjs.isCrossDatasetReferenceSchemaType;
export const isDeleteMutation = cjs.isDeleteMutation;
export const isDeprecatedSchemaType = cjs.isDeprecatedSchemaType;
export const isDeprecationConfiguration = cjs.isDeprecationConfiguration;
export const isDocumentSchemaType = cjs.isDocumentSchemaType;
export const isImage = cjs.isImage;
export const isIndexSegment = cjs.isIndexSegment;
export const isIndexTuple = cjs.isIndexTuple;
export const isKeySegment = cjs.isKeySegment;
export const isKeyedObject = cjs.isKeyedObject;
export const isNumberSchemaType = cjs.isNumberSchemaType;
export const isObjectSchemaType = cjs.isObjectSchemaType;
export const isPatchMutation = cjs.isPatchMutation;
export const isPortableTextListBlock = cjs.isPortableTextListBlock;
export const isPortableTextSpan = cjs.isPortableTextSpan;
export const isPortableTextTextBlock = cjs.isPortableTextTextBlock;
export const isPrimitiveSchemaType = cjs.isPrimitiveSchemaType;
export const isReference = cjs.isReference;
export const isReferenceSchemaType = cjs.isReferenceSchemaType;
export const isSanityDocument = cjs.isSanityDocument;
export const isSlug = cjs.isSlug;
export const isSpanSchemaType = cjs.isSpanSchemaType;
export const isStringSchemaType = cjs.isStringSchemaType;
export const isTitledListValue = cjs.isTitledListValue;
export const isTypedObject = cjs.isTypedObject;
export const isValidationError = cjs.isValidationError;
export const isValidationErrorMarker = cjs.isValidationErrorMarker;
export const isValidationInfo = cjs.isValidationInfo;
export const isValidationInfoMarker = cjs.isValidationInfoMarker;
export const isValidationWarning = cjs.isValidationWarning;
export const isValidationWarningMarker = cjs.isValidationWarningMarker;
export const typed = cjs.typed;
