// @sanity/vision/lib/_chunks/index-DH2XRlCx.js
'use strict';

var icons = require('@sanity/icons');
var react = require('react');
var sanity = require('sanity');
var router = require('sanity/router');
const visionLocaleNamespace = "vision";
const visionUsEnglishLocaleBundle = sanity.defineLocaleResourceBundle({
  locale: "en-US",
  namespace: visionLocaleNamespace,
  resources: () => Promise.resolve().then(function () {
    return require('./resources-ZIFnWdKB.js');
  })
});
const visionTool = sanity.definePlugin(options => {
  const {
    name,
    title,
    icon,
    ...config
  } = options || {};
  return {
    name: "@sanity/vision",
    tools: [{
      name: name || "vision",
      title: title || "Vision",
      icon: icon || icons.EyeOpenIcon,
      component: react.lazy(() => Promise.resolve().then(function () {
        return require('./SanityVision-hdn1_KXE.js');
      })),
      options: config,
      router: router.route.create("/*")
    }],
    i18n: {
      bundles: [visionUsEnglishLocaleBundle]
    }
  };
});
exports.visionLocaleNamespace = visionLocaleNamespace;
exports.visionTool = visionTool;
