// @sanity/vision/lib/_chunks/SanityVision-hdn1_KXE.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var sanity = require('sanity');
var ui = require('@sanity/ui');
var react = require('react');
var SplitPane = require('@rexxars/react-split-pane');
var icons = require('@sanity/icons');
var isHotkey = require('is-hotkey');
var CodeMirror = require('@uiw/react-codemirror');
var autocomplete = require('@codemirror/autocomplete');
var commands = require('@codemirror/commands');
var langJavascript = require('@codemirror/lang-javascript');
var language = require('@codemirror/language');
var search = require('@codemirror/search');
var view = require('@codemirror/view');
var highlight = require('@lezer/highlight');
var color = require('@sanity/color');
var styled = require('styled-components');
var resizeObserver = require('@juggle/resize-observer');
var JSON5 = require('json5');
var debounce = require('lodash/debounce.js');
var index = require('./index-DH2XRlCx.js');
var JSONInspector = require('@rexxars/react-json-inspector');
var HLRU = require('hashlru');
function _interopDefaultCompat(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    default: e
  };
}
var SplitPane__default = /*#__PURE__*/_interopDefaultCompat(SplitPane);
var isHotkey__default = /*#__PURE__*/_interopDefaultCompat(isHotkey);
var CodeMirror__default = /*#__PURE__*/_interopDefaultCompat(CodeMirror);
var styled__default = /*#__PURE__*/_interopDefaultCompat(styled);
var JSON5__default = /*#__PURE__*/_interopDefaultCompat(JSON5);
var debounce__default = /*#__PURE__*/_interopDefaultCompat(debounce);
var JSONInspector__default = /*#__PURE__*/_interopDefaultCompat(JSONInspector);
var HLRU__default = /*#__PURE__*/_interopDefaultCompat(HLRU);
const API_VERSIONS = ["v1", "vX", "v2021-03-25", "v2021-10-21", "v2022-03-07"];
const [DEFAULT_API_VERSION] = API_VERSIONS.slice(-1);
function DelayedSpinner(props) {
  const [show, setShow] = react.useState(false);
  react.useEffect(() => {
    const timer = setTimeout(() => setShow(true), props.delay || 500);
    return () => clearTimeout(timer);
  }, [props.delay]);
  return show ? /* @__PURE__ */jsxRuntime.jsx(ui.Spinner, {
    muted: true,
    size: 4
  }) : null;
}
const codemirrorExtensions = [[langJavascript.javascriptLanguage], view.lineNumbers(), view.highlightActiveLine(), view.highlightActiveLineGutter(), search.highlightSelectionMatches(), view.highlightSpecialChars(), language.indentOnInput(), language.bracketMatching(), autocomplete.closeBrackets(), commands.history(), view.drawSelection(), language.syntaxHighlighting(language.defaultHighlightStyle, {
  fallback: true
}), view.keymap.of([commands.defaultKeymap, commands.historyKeymap].flat().filter(Boolean))];
function useCodemirrorTheme(theme) {
  const cmTheme = react.useMemo(() => createTheme(theme), [theme]);
  const cmHighlight = react.useMemo(() => language.syntaxHighlighting(createHighlight(theme)), [theme]);
  return [cmTheme, cmHighlight];
}
function createTheme(theme) {
  const {
    color: color$1,
    fonts
  } = theme.sanity;
  const card = color$1.card.enabled;
  const cursor = color.hues.blue[color$1.dark ? 400 : 500].hex;
  const selection = color.hues.gray[theme.sanity.color.dark ? 900 : 100].hex;
  return view.EditorView.theme({
    "&": {
      color: card.fg,
      backgroundColor: card.bg
    },
    ".cm-content": {
      caretColor: cursor
    },
    ".cm-editor": {
      fontFamily: fonts.code.family,
      fontSize: ui.rem(fonts.code.sizes[1].fontSize),
      lineHeight: "inherit"
    },
    ".cm-cursor, .cm-dropCursor": {
      borderLeftColor: cursor
    },
    "&.cm-focused .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection": {
      backgroundColor: selection
    },
    ".cm-panels": {
      backgroundColor: card.bg,
      color: card.fg
    },
    ".cm-panels.cm-panels-top": {
      borderBottom: "2px solid ".concat(card.border)
    },
    ".cm-panels.cm-panels-bottom": {
      borderTop: "2px solid ".concat(card.border)
    }
  }, {
    dark: color$1.dark
  });
}
function createHighlight(theme) {
  const c = theme.sanity.color.base;
  const s = theme.sanity.color.syntax;
  return language.HighlightStyle.define([{
    tag: highlight.tags.keyword,
    color: s.keyword
  }, {
    tag: [highlight.tags.propertyName, highlight.tags.name, highlight.tags.deleted, highlight.tags.character, highlight.tags.macroName],
    color: s.property
  }, {
    tag: [highlight.tags.function(highlight.tags.variableName), highlight.tags.labelName],
    color: s.function
  }, {
    tag: [highlight.tags.color, highlight.tags.constant(highlight.tags.name), highlight.tags.standard(highlight.tags.name)],
    color: s.variable
  }, {
    tag: [highlight.tags.definition(highlight.tags.name), highlight.tags.separator],
    color: s.constant
  }, {
    tag: [highlight.tags.typeName, highlight.tags.className, highlight.tags.number, highlight.tags.changed, highlight.tags.annotation, highlight.tags.modifier, highlight.tags.self, highlight.tags.namespace],
    color: s.number
  }, {
    tag: [highlight.tags.operator, highlight.tags.operatorKeyword, highlight.tags.url, highlight.tags.escape, highlight.tags.regexp, highlight.tags.link, highlight.tags.special(highlight.tags.string)],
    color: s.operator
  }, {
    tag: [highlight.tags.meta, highlight.tags.comment],
    color: s.comment
  }, {
    tag: highlight.tags.strong,
    fontWeight: "bold"
  }, {
    tag: highlight.tags.emphasis,
    fontStyle: "italic"
  }, {
    tag: highlight.tags.strikethrough,
    textDecoration: "line-through"
  }, {
    tag: highlight.tags.heading,
    fontWeight: "bold",
    color: s.property
  }, {
    tag: [highlight.tags.atom, highlight.tags.bool, highlight.tags.special(highlight.tags.variableName)],
    color: s.boolean
  }, {
    tag: [highlight.tags.processingInstruction, highlight.tags.string, highlight.tags.inserted],
    color: s.string
  }, {
    tag: highlight.tags.invalid,
    color: c.fg
  }]);
}
var __freeze$4 = Object.freeze;
var __defProp$4 = Object.defineProperty;
var __template$4 = (cooked, raw) => __freeze$4(__defProp$4(cooked, "raw", {
  value: __freeze$4(raw || cooked.slice())
}));
var _a$4;
const EditorRoot = styled__default.default.div(_a$4 || (_a$4 = __template$4(["\n  width: 100%;\n  box-sizing: border-box;\n  height: 100%;\n  overflow: hidden;\n  overflow: clip;\n  position: relative;\n  display: flex;\n\n  & .cm-theme {\n    width: 100%;\n  }\n\n  & .cm-editor {\n    height: 100%;\n\n    font-size: 16px;\n    line-height: 21px;\n  }\n\n  & .cm-line {\n    padding-left: ", ";\n  }\n\n  & .cm-content {\n    border-right-width: ", " !important;\n    padding-top: ", ";\n  }\n"])), _ref => {
  let {
    theme
  } = _ref;
  return ui.rem(theme.sanity.space[3]);
}, _ref2 => {
  let {
    theme
  } = _ref2;
  return ui.rem(theme.sanity.space[4]);
}, _ref3 => {
  let {
    theme
  } = _ref3;
  return ui.rem(theme.sanity.space[5]);
});
function VisionCodeMirror(props) {
  const sanityTheme = ui.useTheme();
  const theme = useCodemirrorTheme(sanityTheme);
  return /* @__PURE__ */jsxRuntime.jsx(EditorRoot, {
    children: /* @__PURE__ */jsxRuntime.jsx(CodeMirror__default.default, {
      basicSetup: false,
      theme,
      extensions: codemirrorExtensions,
      ...props
    })
  });
}
const PERSPECTIVES = ["raw", "previewDrafts", "published"];
const DEFAULT_PERSPECTIVE = PERSPECTIVES[0];
function isPerspective(p) {
  return PERSPECTIVES.includes(p);
}
function encodeQueryString(query) {
  let params = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  let options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  const searchParams = new URLSearchParams();
  searchParams.set("query", query);
  for (const [key, value] of Object.entries(params)) {
    searchParams.set("$".concat(key), JSON.stringify(value));
  }
  for (const [key, value] of Object.entries(options)) {
    if (value) searchParams.set(key, "".concat(value));
  }
  return "?".concat(searchParams);
}
function isPlainObject(obj) {
  return !!obj && typeof obj === "object" && Object.prototype.toString.call(obj) === "[object Object]";
}
const hasLocalStorage = supportsLocalStorage();
const keyPrefix = "sanityVision:";
function clearLocalStorage() {
  if (!hasLocalStorage) {
    return;
  }
  for (let i = 0; i < localStorage.length; i++) {
    const key = localStorage.key(i);
    if (key == null ? void 0 : key.startsWith(keyPrefix)) {
      localStorage.removeItem(key);
    }
  }
}
function getLocalStorage(namespace) {
  const storageKey = "".concat(keyPrefix).concat(namespace);
  let loadedState = null;
  return {
    get,
    set,
    merge
  };
  function get(key, defaultVal) {
    const state = ensureState();
    return typeof state[key] === "undefined" ? defaultVal : state[key];
  }
  function set(key, value) {
    const state = ensureState();
    state[key] = value;
    localStorage.setItem(storageKey, JSON.stringify(loadedState));
    return value;
  }
  function merge(props) {
    const state = {
      ...ensureState(),
      ...props
    };
    localStorage.setItem(storageKey, JSON.stringify(state));
    return state;
  }
  function ensureState() {
    if (loadedState === null) {
      loadedState = loadState();
    }
    return loadedState;
  }
  function loadState() {
    if (!hasLocalStorage) {
      return {};
    }
    try {
      const stored = JSON.parse(localStorage.getItem(storageKey) || "{}");
      return isPlainObject(stored) ? stored : {};
    } catch (err) {
      return {};
    }
  }
}
function supportsLocalStorage() {
  const mod = "lsCheck";
  try {
    localStorage.setItem(mod, mod);
    localStorage.removeItem(mod);
    return true;
  } catch (err) {
    return false;
  }
}
function parseApiQueryString(qs) {
  const params = {};
  const options = {};
  for (const [key, value] of qs.entries()) {
    if (key[0] === "$") {
      params[key.slice(1)] = JSON.parse(value);
      continue;
    }
    if (key === "perspective") {
      options[key] = value;
      continue;
    }
  }
  return {
    query: qs.get("query") || "",
    params,
    options
  };
}
function prefixApiVersion(version) {
  if (version[0] !== "v" && version !== "other") {
    return "v".concat(version);
  }
  return version;
}
const RO = typeof document === "undefined" ? resizeObserver.ResizeObserver : window.ResizeObserver;
function tryParseParams(val, t) {
  try {
    const parsed = val ? JSON5__default.default.parse(val) : {};
    return typeof parsed === "object" && parsed && !Array.isArray(parsed) ? parsed : {};
  } catch (err) {
    err.message = "".concat(t("params.error.params-invalid-json"), ":\n\n").concat(err.message.replace("JSON5:", ""));
    return err;
  }
}
function validateApiVersion(apiVersion) {
  const parseableApiVersion = apiVersion.replace(/^v/, "").trim().toUpperCase();
  const isValidApiVersion = parseableApiVersion.length > 0 && (parseableApiVersion === "X" || parseableApiVersion === "1" || /^\d{4}-\d{2}-\d{2}$/.test(parseableApiVersion) && !isNaN(Date.parse(parseableApiVersion)));
  return isValidApiVersion;
}
const defaultValue = "{\n  \n}";
function ParamsEditor(props) {
  const {
    onChange
  } = props;
  const {
    t
  } = sanity.useTranslation(index.visionLocaleNamespace);
  const {
    raw: value,
    error,
    parsed,
    valid
  } = eventFromValue(props.value, t);
  const [isValid, setValid] = react.useState(valid);
  const [init, setInit] = react.useState(false);
  react.useEffect(() => {
    if (!init) {
      onChange({
        parsed,
        raw: value,
        valid: isValid,
        error
      });
      setInit(true);
    }
  }, [error, init, isValid, onChange, parsed, value]);
  const handleChangeRaw = react.useCallback(newValue => {
    const event = eventFromValue(newValue, t);
    setValid(event.valid);
    onChange(event);
  }, [onChange, t]);
  const handleChange = react.useMemo(() => debounce__default.default(handleChangeRaw, 333), [handleChangeRaw]);
  return /* @__PURE__ */jsxRuntime.jsx(VisionCodeMirror, {
    value: props.value || defaultValue,
    onChange: handleChange
  });
}
function eventFromValue(value, t) {
  const parsedParams = tryParseParams(value, t);
  const params = parsedParams instanceof Error ? {} : parsedParams;
  const validationError = parsedParams instanceof Error ? parsedParams.message : void 0;
  const isValid = !validationError;
  return {
    parsed: params,
    raw: value,
    valid: isValid,
    error: validationError
  };
}
var __freeze$3 = Object.freeze;
var __defProp$3 = Object.defineProperty;
var __template$3 = (cooked, raw) => __freeze$3(__defProp$3(cooked, "raw", {
  value: __freeze$3(raw || cooked.slice())
}));
var _a$3, _b$1;
const PerspectivePopoverContent = styled__default.default(ui.Box)(_a$3 || (_a$3 = __template$3(["\n  /* This limits the width of the popover content */\n  max-width: 240px;\n"])));
const PerspectivePopoverLink = styled__default.default.a(_b$1 || (_b$1 = __template$3(["\n  cursor: pointer;\n  margin-right: auto;\n"])));
function PerspectivePopover() {
  const [open, setOpen] = react.useState(false);
  const [buttonEl, setButtonEl] = react.useState(null);
  const [popoverEl, setPopoverEl] = react.useState(null);
  const handleClick = react.useCallback(() => setOpen(o => !o), []);
  const handleClickOutside = react.useCallback(() => setOpen(false), []);
  const {
    t
  } = sanity.useTranslation(index.visionLocaleNamespace);
  ui.useClickOutside(handleClickOutside, [buttonEl, popoverEl]);
  return /* @__PURE__ */jsxRuntime.jsx(ui.Popover, {
    content: /* @__PURE__ */jsxRuntime.jsx(PerspectivePopoverContent, {
      children: /* @__PURE__ */jsxRuntime.jsxs(ui.Stack, {
        space: 4,
        children: [/* @__PURE__ */jsxRuntime.jsxs(ui.Inline, {
          space: 2,
          children: [/* @__PURE__ */jsxRuntime.jsx(ui.Text, {
            weight: "medium",
            children: t("settings.perspectives.title")
          }), /* @__PURE__ */jsxRuntime.jsx(ui.Badge, {
            tone: "primary",
            children: t("label.new")
          })]
        }), /* @__PURE__ */jsxRuntime.jsx(ui.Card, {
          children: /* @__PURE__ */jsxRuntime.jsx(ui.Text, {
            muted: true,
            children: t("settings.perspectives.description")
          })
        }), /* @__PURE__ */jsxRuntime.jsx(ui.Card, {
          children: /* @__PURE__ */jsxRuntime.jsx(ui.Text, {
            children: /* @__PURE__ */jsxRuntime.jsxs(PerspectivePopoverLink, {
              href: "https://sanity.io/docs/perspectives",
              target: "_blank",
              children: [t("settings.perspectives.action.docs-link"), " \u2192"]
            })
          })
        })]
      })
    }),
    placement: "bottom-start",
    portal: true,
    padding: 3,
    ref: setPopoverEl,
    open,
    children: /* @__PURE__ */jsxRuntime.jsx(ui.Button, {
      icon: icons.HelpCircleIcon,
      mode: "bleed",
      padding: 2,
      tone: "primary",
      fontSize: 1,
      ref: setButtonEl,
      onClick: handleClick,
      selected: open
    })
  });
}
var __freeze$2 = Object.freeze;
var __defProp$2 = Object.defineProperty;
var __template$2 = (cooked, raw) => __freeze$2(__defProp$2(cooked, "raw", {
  value: __freeze$2(raw || cooked.slice())
}));
var _a$2;
const ErrorCode = styled__default.default(ui.Code)(_a$2 || (_a$2 = __template$2(["\n  color: ", ";\n"])), _ref4 => {
  let {
    theme
  } = _ref4;
  return theme.sanity.color.muted.critical.enabled.fg;
});
function QueryErrorDetails(_ref5) {
  let {
    error
  } = _ref5;
  const {
    t
  } = sanity.useTranslation(index.visionLocaleNamespace);
  if (!("details" in error)) {
    return null;
  }
  const details = {
    ...error.details,
    ...mapToLegacyDetails(error.details)
  };
  if (!details.line) {
    return null;
  }
  return /* @__PURE__ */jsxRuntime.jsxs("div", {
    children: [/* @__PURE__ */jsxRuntime.jsx(ErrorCode, {
      size: 1,
      children: "".concat(details.line, "\n").concat(dashLine(details.column, details.columnEnd))
    }), /* @__PURE__ */jsxRuntime.jsx(ui.Box, {
      marginTop: 4,
      children: /* @__PURE__ */jsxRuntime.jsx(ErrorCode, {
        size: 1,
        children: "".concat(t("query.error.line"), ":   ").concat(details.lineNumber, "\n").concat(t("query.error.column"), ": ").concat(details.column)
      })
    })]
  });
}
function mapToLegacyDetails(details) {
  if (!details || typeof details.query !== "string" || typeof details.start !== "number") {
    return {};
  }
  const {
    query,
    start,
    end
  } = details;
  const lineStart = query.slice(0, start).lastIndexOf("\n") + 1;
  const lineNumber = (query.slice(0, lineStart).match(/\n/g) || []).length;
  const line = query.slice(lineStart, query.indexOf("\n", lineStart));
  const column = start - lineStart;
  const columnEnd = typeof end === "number" ? end - lineStart : void 0;
  return {
    line,
    lineNumber,
    column,
    columnEnd
  };
}
function dashLine(column, columnEnd) {
  const line = "-".repeat(column);
  const hats = "^".repeat(columnEnd ? columnEnd - column : 1);
  return "".concat(line).concat(hats);
}
function QueryErrorDialog(props) {
  return /* @__PURE__ */jsxRuntime.jsxs(ui.Stack, {
    space: 5,
    marginTop: 2,
    children: [/* @__PURE__ */jsxRuntime.jsx(ErrorCode, {
      size: 1,
      children: props.error.message
    }), /* @__PURE__ */jsxRuntime.jsx(QueryErrorDetails, {
      error: props.error
    })]
  });
}
var __freeze$1 = Object.freeze;
var __defProp$1 = Object.defineProperty;
var __template$1 = (cooked, raw) => __freeze$1(__defProp$1(cooked, "raw", {
  value: __freeze$1(raw || cooked.slice())
}));
var _a$1;
const ResultViewWrapper = styled__default.default.div(_ref6 => {
  let {
    theme
  } = _ref6;
  const {
    color,
    fonts,
    space
  } = theme.sanity;
  return styled.css(_a$1 || (_a$1 = __template$1(["\n    & .json-inspector,\n    & .json-inspector .json-inspector__selection {\n      font-family: ", ";\n      font-size: ", "px;\n      line-height: ", "px;\n      color: var(--card-code-fg-color);\n    }\n\n    & .json-inspector .json-inspector__leaf {\n      padding-left: ", ";\n    }\n\n    & .json-inspector .json-inspector__leaf.json-inspector__leaf_root {\n      padding-top: ", ";\n      padding-left: 0;\n    }\n\n    & .json-inspector > .json-inspector__leaf_root > .json-inspector__line > .json-inspector__key {\n      display: none;\n    }\n\n    & .json-inspector .json-inspector__line {\n      display: block;\n      position: relative;\n      cursor: default;\n    }\n\n    & .json-inspector .json-inspector__line::after {\n      content: '';\n      position: absolute;\n      top: 0;\n      left: -200px;\n      right: -50px;\n      bottom: 0;\n      z-index: -1;\n      pointer-events: none;\n    }\n\n    & .json-inspector .json-inspector__line:hover::after {\n      background: var(--card-code-bg-color);\n    }\n\n    & .json-inspector .json-inspector__leaf_composite > .json-inspector__line {\n      cursor: pointer;\n    }\n\n    & .json-inspector .json-inspector__leaf_composite > .json-inspector__line::before {\n      content: '\u25B8 ';\n      margin-left: calc(0px - ", ");\n      font-size: ", "px;\n      line-height: ", "px;\n    }\n\n    &\n      .json-inspector\n      .json-inspector__leaf_expanded.json-inspector__leaf_composite\n      > .json-inspector__line::before {\n      content: '\u25BE ';\n      font-size: ", "px;\n      line-height: ", "px;\n    }\n\n    & .json-inspector .json-inspector__radio,\n    & .json-inspector .json-inspector__flatpath {\n      display: none;\n    }\n\n    & .json-inspector .json-inspector__value {\n      margin-left: ", ";\n    }\n\n    &\n      .json-inspector\n      > .json-inspector__leaf_root\n      > .json-inspector__line\n      > .json-inspector__key\n      + .json-inspector__value {\n      margin: 0;\n    }\n\n    & .json-inspector .json-inspector__key {\n      color: ", ";\n    }\n\n    & .json-inspector .json-inspector__value_helper,\n    & .json-inspector .json-inspector__value_null {\n      color: ", ";\n    }\n\n    & .json-inspector .json-inspector__not-found {\n      padding-top: ", ";\n    }\n\n    & .json-inspector .json-inspector__value_string {\n      color: ", ";\n      word-break: break-word;\n    }\n\n    & .json-inspector .json-inspector__value_boolean {\n      color: ", ";\n    }\n\n    & .json-inspector .json-inspector__value_number {\n      color: ", ";\n    }\n\n    & .json-inspector .json-inspector__show-original {\n      display: inline-block;\n      padding: 0 6px;\n      cursor: pointer;\n    }\n\n    & .json-inspector .json-inspector__show-original:hover {\n      color: inherit;\n    }\n\n    & .json-inspector .json-inspector__show-original::before {\n      content: '\u2194';\n    }\n\n    & .json-inspector .json-inspector__show-original:hover::after {\n      content: ' expand';\n    }\n  "])), fonts.code.family, fonts.code.sizes[2].fontSize, fonts.code.sizes[2].lineHeight, ui.rem(space[4]), ui.rem(space[0]), ui.rem(space[4]), fonts.code.sizes[2].fontSize, fonts.code.sizes[2].lineHeight, fonts.code.sizes[2].fontSize, fonts.code.sizes[2].lineHeight, ui.rem(space[4] / 2), color.syntax.property, color.syntax.constant, ui.rem(space[2]), color.syntax.string, color.syntax.boolean, color.syntax.number);
});
const lru = HLRU__default.default(5e4);
function ResultView(props) {
  const {
    data
  } = props;
  if (isRecord(data) || Array.isArray(data)) {
    return /* @__PURE__ */jsxRuntime.jsx(ResultViewWrapper, {
      children: /* @__PURE__ */jsxRuntime.jsx(JSONInspector__default.default, {
        data,
        search: false,
        isExpanded,
        onClick: toggleExpanded
      })
    });
  }
  return /* @__PURE__ */jsxRuntime.jsx(ui.Code, {
    language: "json",
    children: JSON.stringify(data)
  });
}
function isExpanded(keyPath, value) {
  const depthLimit = 4;
  const cached = lru.get(keyPath);
  if (typeof cached === "boolean") {
    return cached;
  }
  const segments = keyPath.split(".", depthLimit);
  if (segments.length === depthLimit) {
    return false;
  }
  if (Array.isArray(value)) {
    return true;
  }
  return isRecord(value) && !segments.some(key => isArrayKeyOverLimit(key));
}
function toggleExpanded(event) {
  const {
    path
  } = event;
  const current = lru.get(path);
  if (current === void 0) {
    return;
  }
  lru.set(path, !current);
}
function isRecord(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}
const numeric = /^\d+$/;
function isArrayKeyOverLimit(segment) {
  let limit = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 10;
  return numeric.test(segment) && parseInt(segment, 10) > limit;
}
var __freeze = Object.freeze;
var __defProp = Object.defineProperty;
var __template = (cooked, raw) => __freeze(__defProp(cooked, "raw", {
  value: __freeze(raw || cooked.slice())
}));
var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l, _m, _n, _o, _p, _q, _r;
const Root = styled__default.default(ui.Flex)(_a || (_a = __template(["\n  .sidebarPanes .Pane {\n    overflow-y: auto;\n    overflow-x: hidden;\n  }\n\n  & .Resizer {\n    background: var(--card-border-color);\n    opacity: 1;\n    z-index: 1;\n    box-sizing: border-box;\n    background-clip: padding-box;\n    border: solid transparent;\n  }\n\n  & .Resizer:hover {\n    border-color: var(--card-shadow-ambient-color);\n  }\n\n  & .Resizer.horizontal {\n    height: 11px;\n    margin: -5px 0;\n    border-width: 5px 0;\n    cursor: row-resize;\n    width: 100%;\n    z-index: 4;\n  }\n\n  & .Resizer.vertical {\n    width: 11px;\n    margin: 0 -5px;\n    border-width: 0 5px;\n    cursor: col-resize;\n    z-index: 2; /* To prevent the resizer from being hidden behind CodeMirror scroll area */\n  }\n\n  .Resizer.disabled {\n    cursor: not-allowed;\n  }\n\n  .Resizer.disabled:hover {\n    border-color: transparent;\n  }\n"])));
Root.displayName = "Root";
const Header = styled__default.default(ui.Card)(_b || (_b = __template(["\n  border-bottom: 1px solid var(--card-border-color);\n"])));
const StyledLabel = styled__default.default(ui.Label)(_c || (_c = __template(["\n  flex: 1;\n"])));
const SplitpaneContainer = styled__default.default(ui.Box)(_d || (_d = __template(["\n  position: relative;\n"])));
const QueryCopyLink = styled__default.default.a(_e || (_e = __template(["\n  cursor: pointer;\n  margin-right: auto;\n"])));
const InputBackgroundContainer = styled__default.default(ui.Box)(_f || (_f = __template(["\n  position: absolute;\n  top: 1rem;\n  left: 0;\n  padding: 0;\n  margin: 0;\n  z-index: 10;\n  right: 0;\n\n  ", " {\n    user-select: none;\n  }\n"])), StyledLabel);
const InputBackgroundContainerLeft = styled__default.default(InputBackgroundContainer)(_g || (_g = __template(["\n  // This is so its aligned with the gutters of CodeMirror\n  left: 33px;\n"])));
const InputContainer = styled__default.default(ui.Card)(_h || (_h = __template(["\n  width: 100%;\n  height: 100%;\n  position: relative;\n  flex-direction: column;\n"])));
const ResultOuterContainer = styled__default.default(ui.Flex)(_i || (_i = __template(["\n  height: 100%;\n"])));
const ResultInnerContainer = styled__default.default(ui.Box)(_j || (_j = __template(["\n  position: relative;\n"])));
const ResultContainer = styled__default.default(ui.Card)(_l || (_l = __template(["\n  height: 100%;\n  width: 100%;\n  position: absolute;\n  max-width: 100%;\n\n  ", "\n"])), _ref7 => {
  let {
    $isInvalid
  } = _ref7;
  return $isInvalid && styled.css(_k || (_k = __template(["\n      &:after {\n        background-color: var(--card-bg-color);\n        content: '';\n        position: absolute;\n        top: 0;\n        bottom: 0;\n        left: 0;\n        width: 100%;\n      }\n    "])));
});
const Result = styled__default.default(ui.Box)(_m || (_m = __template(["\n  position: relative;\n  width: 100%;\n  height: 100%;\n  z-index: 20;\n"])));
const TimingsFooter = styled__default.default(ui.Box)(_n || (_n = __template(["\n  border-top: 1px solid var(--card-border-color);\n"])));
const TimingsCard = styled__default.default(ui.Card)(_o || (_o = __template(["\n  position: relative;\n"])));
styled__default.default(ui.Box)(_p || (_p = __template(["\n  width: 100%;\n  height: 100%;\n"])));
const TimingsTextContainer = styled__default.default(ui.Flex)(_q || (_q = __template(["\n  height: 100%;\n  min-height: ", ";\n"])), _ref8 => {
  let {
    theme
  } = _ref8;
  return ui.rem(theme.sanity.space[3] * 2 + theme.sanity.fonts.text.sizes[2].lineHeight - theme.sanity.fonts.text.sizes[2].ascenderHeight - theme.sanity.fonts.text.sizes[2].descenderHeight);
});
const ControlsContainer = styled__default.default(ui.Box)(_r || (_r = __template(["\n  border-top: 1px solid var(--card-border-color);\n"])));
function nodeContains(node, other) {
  if (!node || !other) {
    return false;
  }
  return node === other || !!(node.compareDocumentPosition(other) & 16);
}
const sanityUrl = /\.(?:api|apicdn)\.sanity\.io\/(vX|v1|v\d{4}-\d\d-\d\d)\/.*?(?:query|listen)\/(.*?)\?(.*)/;
const isRunHotkey = event => isHotkey__default.default("ctrl+enter", event) || isHotkey__default.default("mod+enter", event);
function narrowBreakpoint() {
  return typeof window !== "undefined" && window.innerWidth > 600;
}
function calculatePaneSizeOptions(rootHeight) {
  return {
    defaultSize: rootHeight / (narrowBreakpoint() ? 2 : 1),
    size: rootHeight > 550 ? void 0 : rootHeight * 0.4,
    allowResize: rootHeight > 550,
    minSize: Math.min(170, Math.max(170, rootHeight / 2)),
    maxSize: rootHeight > 650 ? rootHeight * 0.7 : rootHeight * 0.6
  };
}
class VisionGui extends react.PureComponent {
  constructor(props) {
    super(props);
    const {
      client,
      datasets,
      config
    } = props;
    this._localStorage = getLocalStorage(client.config().projectId || "default");
    const defaultDataset = config.defaultDataset || client.config().dataset || datasets[0];
    const defaultApiVersion = prefixApiVersion("".concat(config.defaultApiVersion));
    const defaultPerspective = DEFAULT_PERSPECTIVE;
    let dataset = this._localStorage.get("dataset", defaultDataset);
    let apiVersion = this._localStorage.get("apiVersion", defaultApiVersion);
    let lastQuery = this._localStorage.get("query", "");
    let lastParams = this._localStorage.get("params", "{\n  \n}");
    const customApiVersion = API_VERSIONS.includes(apiVersion) ? false : apiVersion;
    let perspective = this._localStorage.get("perspective", defaultPerspective);
    if (!datasets.includes(dataset)) {
      dataset = datasets.includes(defaultDataset) ? defaultDataset : datasets[0];
    }
    if (!API_VERSIONS.includes(apiVersion)) {
      apiVersion = DEFAULT_API_VERSION;
    }
    if (!PERSPECTIVES.includes(perspective)) {
      perspective = DEFAULT_PERSPECTIVE;
    }
    if (typeof lastQuery !== "string") {
      lastQuery = "";
    }
    if (typeof lastParams !== "string") {
      lastParams = "{\n  \n}";
    }
    this._visionRoot = react.createRef();
    this._operationUrlElement = react.createRef();
    this._queryEditorContainer = react.createRef();
    this._paramsEditorContainer = react.createRef();
    this._customApiVersionElement = react.createRef();
    this._client = props.client.withConfig({
      apiVersion: customApiVersion || apiVersion,
      dataset,
      perspective,
      allowReconfigure: true
    });
    const bodyHeight = typeof window !== "undefined" && typeof document !== "undefined" ? document.body.getBoundingClientRect().height - 60 : 0;
    const params = lastParams ? tryParseParams(lastParams, this.props.t) : void 0;
    this.state = {
      // Selected options
      dataset,
      apiVersion,
      customApiVersion,
      perspective,
      // Selected options validation state
      isValidApiVersion: customApiVersion ? validateApiVersion(customApiVersion) : false,
      // Inputs
      query: lastQuery,
      rawParams: lastParams,
      // Parsed input
      params,
      hasValidParams: !(params instanceof Error),
      // Query/listen results
      listenMutations: [],
      // Operation state
      queryInProgress: false,
      listenInProgress: false,
      // UI drawing state
      paneSizeOptions: calculatePaneSizeOptions(bodyHeight)
    };
    this.handleChangeDataset = this.handleChangeDataset.bind(this);
    this.handleChangeApiVersion = this.handleChangeApiVersion.bind(this);
    this.handleCustomApiVersionChange = this.handleCustomApiVersionChange.bind(this);
    this.handleChangePerspective = this.handleChangePerspective.bind(this);
    this.handleListenExecution = this.handleListenExecution.bind(this);
    this.handleListenerEvent = this.handleListenerEvent.bind(this);
    this.handleQueryExecution = this.handleQueryExecution.bind(this);
    this.handleQueryChange = this.handleQueryChange.bind(this);
    this.handleParamsChange = this.handleParamsChange.bind(this);
    this.handleCopyUrl = this.handleCopyUrl.bind(this);
    this.handlePaste = this.handlePaste.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleResize = this.handleResize.bind(this);
  }
  componentDidMount() {
    window.document.addEventListener("paste", this.handlePaste);
    window.document.addEventListener("keydown", this.handleKeyDown);
    this.handleResizeListen();
  }
  componentWillUnmount() {
    this.cancelQuery();
    this.cancelListener();
    this.cancelEventListener();
    this.cancelResizeListener();
  }
  handleResizeListen() {
    if (!this._visionRoot.current) {
      return;
    }
    this._resizeListener = new RO(this.handleResize);
    this._resizeListener.observe(this._visionRoot.current);
  }
  handleResize(entries) {
    const entry = entries == null ? void 0 : entries[0];
    this.setState(prevState => ({
      ...prevState,
      paneSizeOptions: calculatePaneSizeOptions(entry.contentRect.height)
    }));
  }
  cancelResizeListener() {
    if (this._resizeListener) {
      this._resizeListener.disconnect();
    }
  }
  handlePaste(evt) {
    if (!evt.clipboardData) {
      return;
    }
    const data = evt.clipboardData.getData("text/plain");
    const match = data.match(sanityUrl);
    if (!match) {
      return;
    }
    const [, usedApiVersion, usedDataset, urlQuery] = match;
    let parts;
    try {
      const qs = new URLSearchParams(urlQuery);
      parts = parseApiQueryString(qs);
    } catch (err) {
      console.warn("Error while trying to parse API URL: ", err.message);
      return;
    }
    let apiVersion;
    let customApiVersion;
    if (validateApiVersion(usedApiVersion)) {
      if (API_VERSIONS.includes(usedApiVersion)) {
        apiVersion = usedApiVersion;
        customApiVersion = false;
      } else {
        customApiVersion = usedApiVersion;
      }
    }
    const perspective = PERSPECTIVES.includes(parts.options.perspective) ? parts.options.perspective : void 0;
    evt.preventDefault();
    this.setState(prevState => ({
      dataset: this.props.datasets.includes(usedDataset) ? usedDataset : prevState.dataset,
      query: parts.query,
      params: parts.params,
      rawParams: JSON.stringify(parts.params, null, 2),
      apiVersion: typeof apiVersion === "undefined" ? prevState.apiVersion : apiVersion,
      customApiVersion: typeof customApiVersion === "undefined" ? prevState.customApiVersion : customApiVersion,
      perspective: typeof perspective === "undefined" ? prevState.perspective : perspective
    }), () => {
      this._localStorage.merge({
        query: this.state.query,
        params: this.state.rawParams,
        dataset: this.state.dataset,
        apiVersion: customApiVersion || apiVersion,
        perspective: this.state.perspective
      });
      this._client.config({
        dataset: this.state.dataset,
        apiVersion: customApiVersion || apiVersion,
        perspective: this.state.perspective
      });
      this.handleQueryExecution();
      this.props.toast.push({
        closable: true,
        id: "vision-paste",
        status: "info",
        title: "Parsed URL to query"
      });
    });
  }
  cancelQuery() {
    if (!this._querySubscription) {
      return;
    }
    this._querySubscription.unsubscribe();
    this._querySubscription = void 0;
  }
  cancelListener() {
    if (!this._listenSubscription) {
      return;
    }
    this._listenSubscription.unsubscribe();
    this._listenSubscription = void 0;
  }
  cancelEventListener() {
    window.removeEventListener("keydown", this.handleKeyDown);
  }
  handleChangeDataset(evt) {
    const dataset = evt.target.value;
    this._localStorage.set("dataset", dataset);
    this.setState({
      dataset
    });
    this._client.config({
      dataset
    });
    this.handleQueryExecution();
  }
  handleChangeApiVersion(evt) {
    const apiVersion = evt.target.value;
    if ((apiVersion == null ? void 0 : apiVersion.toLowerCase()) === "other") {
      this.setState({
        customApiVersion: "v"
      }, () => {
        var _a;
        (_a = this._customApiVersionElement.current) == null ? void 0 : _a.focus();
      });
      return;
    }
    this.setState({
      apiVersion,
      customApiVersion: false
    }, () => {
      this._localStorage.set("apiVersion", this.state.apiVersion);
      this._client.config({
        apiVersion: this.state.apiVersion
      });
      this.handleQueryExecution();
    });
  }
  handleCustomApiVersionChange(evt) {
    const customApiVersion = evt.target.value || "";
    const isValidApiVersion = validateApiVersion(customApiVersion);
    this.setState(prevState => ({
      apiVersion: isValidApiVersion ? customApiVersion : prevState.apiVersion,
      customApiVersion: customApiVersion || "v",
      isValidApiVersion
    }), () => {
      if (!this.state.isValidApiVersion || typeof this.state.customApiVersion !== "string") {
        return;
      }
      this._localStorage.set("apiVersion", this.state.customApiVersion);
      this._client.config({
        apiVersion: this.state.customApiVersion
      });
    });
  }
  handleChangePerspective(evt) {
    const perspective = evt.target.value;
    if (!isPerspective(perspective)) {
      return;
    }
    this.setState({
      perspective
    }, () => {
      this._localStorage.set("perspective", this.state.perspective);
      this._client.config({
        perspective: this.state.perspective
      });
      this.handleQueryExecution();
    });
  }
  handleListenerEvent(evt) {
    if (evt.type !== "mutation") {
      this.props.toast.push({
        closable: true,
        id: "vision-listen",
        status: "success",
        title: "Listening for mutations\u2026"
      });
      return;
    }
    this.setState(_ref9 => {
      let {
        listenMutations
      } = _ref9;
      return {
        listenMutations: listenMutations.length === 50 ? [evt, ...listenMutations.slice(0, 49)] : [evt, ...listenMutations]
      };
    });
  }
  handleKeyDown(event) {
    const {
      hasValidParams
    } = this.state;
    const isWithinRoot = this._visionRoot.current && nodeContains(this._visionRoot.current, event.target);
    if (isRunHotkey(event) && isWithinRoot && hasValidParams) {
      this.handleQueryExecution();
      event.preventDefault();
      event.stopPropagation();
    }
  }
  ensureSelectedApiVersion() {
    const {
      apiVersion,
      customApiVersion
    } = this.state;
    const wantedApiVersion = customApiVersion || apiVersion;
    if (this._client.config().apiVersion !== wantedApiVersion) {
      this._client.config({
        apiVersion: wantedApiVersion
      });
    }
  }
  handleListenExecution() {
    const {
      query,
      params,
      rawParams,
      listenInProgress
    } = this.state;
    if (listenInProgress) {
      this.cancelListener();
      this.setState({
        listenInProgress: false
      });
      return;
    }
    this.ensureSelectedApiVersion();
    const paramsError = params instanceof Error ? params : void 0;
    const encodeParams = params instanceof Error ? {} : params || {};
    const url = this._client.getDataUrl("listen", encodeQueryString(query, encodeParams, {}));
    const shouldExecute = !paramsError && query.trim().length > 0;
    this._localStorage.set("query", query);
    this._localStorage.set("params", rawParams);
    this.cancelQuery();
    this.setState({
      url,
      listenMutations: [],
      queryInProgress: false,
      queryResult: void 0,
      listenInProgress: shouldExecute,
      error: paramsError,
      queryTime: void 0,
      e2eTime: void 0
    });
    if (!shouldExecute) {
      return;
    }
    this._listenSubscription = this._client.listen(query, params, {
      events: ["mutation", "welcome"]
    }).subscribe({
      next: this.handleListenerEvent,
      error: error => this.setState({
        error,
        query,
        listenInProgress: false
      })
    });
  }
  handleQueryExecution() {
    const {
      query,
      params,
      rawParams,
      queryInProgress
    } = this.state;
    if (queryInProgress) {
      this.cancelQuery();
      this.cancelListener();
      this.setState({
        queryInProgress: false
      });
      return true;
    }
    const paramsError = params instanceof Error && params;
    this._localStorage.set("query", query);
    this._localStorage.set("params", rawParams);
    this.cancelListener();
    this.setState({
      queryInProgress: !paramsError && Boolean(query),
      listenInProgress: false,
      listenMutations: [],
      error: paramsError || void 0,
      queryResult: void 0,
      queryTime: void 0,
      e2eTime: void 0
    });
    if (!query || paramsError) {
      return true;
    }
    this.ensureSelectedApiVersion();
    const urlQueryOpts = {};
    if (this.state.perspective !== "raw") {
      urlQueryOpts.perspective = this.state.perspective;
    }
    const url = this._client.getUrl(this._client.getDataUrl("query", encodeQueryString(query, params, urlQueryOpts)));
    this.setState({
      url
    });
    const queryStart = Date.now();
    this._querySubscription = this._client.observable.fetch(query, params, {
      filterResponse: false,
      tag: "vision"
    }).subscribe({
      next: res => this.setState({
        queryTime: res.ms,
        e2eTime: Date.now() - queryStart,
        queryResult: res.result,
        queryInProgress: false,
        error: void 0
      }),
      error: error => this.setState({
        error,
        query,
        queryInProgress: false
      })
    });
    return true;
  }
  handleQueryChange(query) {
    this.setState({
      query
    });
  }
  handleParamsChange(_ref10) {
    let {
      raw,
      parsed,
      valid,
      error
    } = _ref10;
    this.setState({
      rawParams: raw,
      params: parsed,
      hasValidParams: valid,
      paramsError: error
    }, () => this._localStorage.set("params", raw));
  }
  handleCopyUrl() {
    const el = this._operationUrlElement.current;
    if (!el) {
      return;
    }
    try {
      el.select();
      document.execCommand("copy");
      this.props.toast.push({
        closable: true,
        title: "Copied to clipboard",
        status: "info",
        id: "vision-copy"
      });
    } catch (err) {
      console.error("Unable to copy to clipboard :(");
    }
  }
  render() {
    const {
      datasets,
      t
    } = this.props;
    const {
      apiVersion,
      customApiVersion,
      dataset,
      e2eTime,
      error,
      hasValidParams,
      isValidApiVersion,
      listenInProgress,
      listenMutations,
      paneSizeOptions,
      paramsError,
      perspective,
      query,
      queryInProgress,
      queryResult,
      queryTime,
      rawParams,
      url
    } = this.state;
    const hasResult = !error && !queryInProgress && typeof queryResult !== "undefined";
    return /* @__PURE__ */jsxRuntime.jsxs(Root, {
      direction: "column",
      height: "fill",
      ref: this._visionRoot,
      sizing: "border",
      overflow: "hidden",
      children: [/* @__PURE__ */jsxRuntime.jsx(Header, {
        paddingX: 3,
        paddingY: 2,
        children: /* @__PURE__ */jsxRuntime.jsxs(ui.Grid, {
          columns: [1, 4, 8, 12],
          children: [/* @__PURE__ */jsxRuntime.jsx(ui.Box, {
            padding: 1,
            column: 2,
            children: /* @__PURE__ */jsxRuntime.jsxs(ui.Stack, {
              children: [/* @__PURE__ */jsxRuntime.jsx(ui.Card, {
                paddingTop: 2,
                paddingBottom: 3,
                children: /* @__PURE__ */jsxRuntime.jsx(StyledLabel, {
                  children: t("settings.dataset-label")
                })
              }), /* @__PURE__ */jsxRuntime.jsx(ui.Select, {
                value: dataset,
                onChange: this.handleChangeDataset,
                children: datasets.map(ds => /* @__PURE__ */jsxRuntime.jsx("option", {
                  children: ds
                }, ds))
              })]
            })
          }), /* @__PURE__ */jsxRuntime.jsx(ui.Box, {
            padding: 1,
            column: 2,
            children: /* @__PURE__ */jsxRuntime.jsxs(ui.Stack, {
              children: [/* @__PURE__ */jsxRuntime.jsx(ui.Card, {
                paddingTop: 2,
                paddingBottom: 3,
                children: /* @__PURE__ */jsxRuntime.jsx(StyledLabel, {
                  children: t("settings.api-version-label")
                })
              }), /* @__PURE__ */jsxRuntime.jsxs(ui.Select, {
                value: customApiVersion === false ? apiVersion : "other",
                onChange: this.handleChangeApiVersion,
                children: [API_VERSIONS.map(version => /* @__PURE__ */jsxRuntime.jsx("option", {
                  children: version
                }, version)), /* @__PURE__ */jsxRuntime.jsx("option", {
                  value: "other",
                  children: t("settings.other-api-version-label")
                }, "other")]
              })]
            })
          }), customApiVersion !== false && /* @__PURE__ */jsxRuntime.jsx(ui.Box, {
            padding: 1,
            column: 2,
            children: /* @__PURE__ */jsxRuntime.jsxs(ui.Stack, {
              children: [/* @__PURE__ */jsxRuntime.jsx(ui.Card, {
                paddingTop: 2,
                paddingBottom: 3,
                children: /* @__PURE__ */jsxRuntime.jsx(StyledLabel, {
                  textOverflow: "ellipsis",
                  children: t("settings.custom-api-version-label")
                })
              }), /* @__PURE__ */jsxRuntime.jsx(ui.TextInput, {
                ref: this._customApiVersionElement,
                value: customApiVersion,
                onChange: this.handleCustomApiVersionChange,
                customValidity: isValidApiVersion ? void 0 : t("settings.error.invalid-api-version"),
                maxLength: 11
              })]
            })
          }), /* @__PURE__ */jsxRuntime.jsx(ui.Box, {
            padding: 1,
            column: 2,
            children: /* @__PURE__ */jsxRuntime.jsxs(ui.Stack, {
              children: [/* @__PURE__ */jsxRuntime.jsx(ui.Card, {
                paddingBottom: 1,
                children: /* @__PURE__ */jsxRuntime.jsxs(ui.Inline, {
                  space: 1,
                  children: [/* @__PURE__ */jsxRuntime.jsx(ui.Box, {
                    children: /* @__PURE__ */jsxRuntime.jsx(StyledLabel, {
                      children: t("settings.perspective-label")
                    })
                  }), /* @__PURE__ */jsxRuntime.jsx(ui.Box, {
                    children: /* @__PURE__ */jsxRuntime.jsx(PerspectivePopover, {})
                  })]
                })
              }), /* @__PURE__ */jsxRuntime.jsx(ui.Select, {
                value: perspective,
                onChange: this.handleChangePerspective,
                children: PERSPECTIVES.map(p => /* @__PURE__ */jsxRuntime.jsx("option", {
                  children: p
                }, p))
              })]
            })
          }), typeof url === "string" ? /* @__PURE__ */jsxRuntime.jsx(ui.Box, {
            padding: 1,
            flex: 1,
            column: customApiVersion === false ? 6 : 4,
            children: /* @__PURE__ */jsxRuntime.jsxs(ui.Stack, {
              children: [/* @__PURE__ */jsxRuntime.jsx(ui.Card, {
                paddingTop: 2,
                paddingBottom: 3,
                children: /* @__PURE__ */jsxRuntime.jsxs(StyledLabel, {
                  children: [t("query.url"), "\xA0", /* @__PURE__ */jsxRuntime.jsxs(QueryCopyLink, {
                    onClick: this.handleCopyUrl,
                    children: ["[", t("action.copy-url-to-clipboard"), "]"]
                  })]
                })
              }), /* @__PURE__ */jsxRuntime.jsxs(ui.Flex, {
                flex: 1,
                gap: 1,
                children: [/* @__PURE__ */jsxRuntime.jsx(ui.Box, {
                  flex: 1,
                  children: /* @__PURE__ */jsxRuntime.jsx(ui.TextInput, {
                    readOnly: true,
                    type: "url",
                    ref: this._operationUrlElement,
                    value: url
                  })
                }), /* @__PURE__ */jsxRuntime.jsx(ui.Tooltip, {
                  content: t("action.copy-url-to-clipboard"),
                  children: /* @__PURE__ */jsxRuntime.jsx(ui.Button, {
                    "aria-label": t("action.copy-url-to-clipboard"),
                    type: "button",
                    mode: "ghost",
                    icon: icons.CopyIcon,
                    onClick: this.handleCopyUrl
                  })
                })]
              })]
            })
          }) : /* @__PURE__ */jsxRuntime.jsx(ui.Box, {
            flex: 1
          })]
        })
      }), /* @__PURE__ */jsxRuntime.jsx(SplitpaneContainer, {
        flex: "auto",
        children: /* @__PURE__ */jsxRuntime.jsxs(SplitPane__default.default, {
          split: narrowBreakpoint() ? "vertical" : "horizontal",
          minSize: 280,
          defaultSize: 400,
          maxSize: -400,
          children: [/* @__PURE__ */jsxRuntime.jsx(ui.Box, {
            height: "stretch",
            flex: 1,
            children: /* @__PURE__ */jsxRuntime.jsxs(SplitPane__default.default, {
              className: "sidebarPanes",
              split: "horizontal",
              defaultSize: narrowBreakpoint() ? paneSizeOptions.defaultSize : paneSizeOptions.minSize,
              size: paneSizeOptions.size,
              allowResize: paneSizeOptions.allowResize,
              minSize: narrowBreakpoint() ? paneSizeOptions.minSize : 100,
              maxSize: paneSizeOptions.maxSize,
              primary: "first",
              children: [/* @__PURE__ */jsxRuntime.jsx(InputContainer, {
                display: "flex",
                ref: this._queryEditorContainer,
                children: /* @__PURE__ */jsxRuntime.jsxs(ui.Box, {
                  flex: 1,
                  children: [/* @__PURE__ */jsxRuntime.jsx(InputBackgroundContainerLeft, {
                    children: /* @__PURE__ */jsxRuntime.jsx(ui.Flex, {
                      children: /* @__PURE__ */jsxRuntime.jsx(StyledLabel, {
                        muted: true,
                        children: t("query.label")
                      })
                    })
                  }), /* @__PURE__ */jsxRuntime.jsx(VisionCodeMirror, {
                    value: query,
                    onChange: this.handleQueryChange
                  })]
                })
              }), /* @__PURE__ */jsxRuntime.jsxs(InputContainer, {
                display: "flex",
                ref: this._paramsEditorContainer,
                children: [/* @__PURE__ */jsxRuntime.jsxs(ui.Card, {
                  flex: 1,
                  tone: hasValidParams ? "default" : "critical",
                  children: [/* @__PURE__ */jsxRuntime.jsx(InputBackgroundContainerLeft, {
                    children: /* @__PURE__ */jsxRuntime.jsxs(ui.Flex, {
                      children: [/* @__PURE__ */jsxRuntime.jsx(StyledLabel, {
                        muted: true,
                        children: t("params.label")
                      }), paramsError && /* @__PURE__ */jsxRuntime.jsx(ui.Tooltip, {
                        placement: "top-end",
                        portal: true,
                        content: paramsError,
                        children: /* @__PURE__ */jsxRuntime.jsx(ui.Box, {
                          padding: 1,
                          marginX: 2,
                          children: /* @__PURE__ */jsxRuntime.jsx(ui.Text, {
                            children: /* @__PURE__ */jsxRuntime.jsx(icons.ErrorOutlineIcon, {})
                          })
                        })
                      })]
                    })
                  }), /* @__PURE__ */jsxRuntime.jsx(ParamsEditor, {
                    value: rawParams,
                    onChange: this.handleParamsChange
                  })]
                }), /* @__PURE__ */jsxRuntime.jsx(ControlsContainer, {
                  children: /* @__PURE__ */jsxRuntime.jsx(ui.Card, {
                    padding: 3,
                    paddingX: 3,
                    children: /* @__PURE__ */jsxRuntime.jsx(ui.Tooltip, {
                      content: /* @__PURE__ */jsxRuntime.jsx(ui.Card, {
                        radius: 4,
                        children: /* @__PURE__ */jsxRuntime.jsx(ui.Text, {
                          size: 1,
                          muted: true,
                          children: t("params.error.params-invalid-json")
                        })
                      }),
                      placement: "top",
                      disabled: hasValidParams,
                      portal: true,
                      children: /* @__PURE__ */jsxRuntime.jsxs(ui.Flex, {
                        justify: "space-evenly",
                        children: [/* @__PURE__ */jsxRuntime.jsx(ui.Box, {
                          flex: 1,
                          children: /* @__PURE__ */jsxRuntime.jsx(ui.Tooltip, {
                            content: /* @__PURE__ */jsxRuntime.jsx(ui.Card, {
                              radius: 4,
                              children: /* @__PURE__ */jsxRuntime.jsx(ui.Hotkeys, {
                                keys: ["Ctrl", "Enter"]
                              })
                            }),
                            placement: "top",
                            portal: true,
                            children: /* @__PURE__ */jsxRuntime.jsx(ui.Button, {
                              width: "fill",
                              onClick: this.handleQueryExecution,
                              type: "button",
                              icon: queryInProgress ? icons.StopIcon : icons.PlayIcon,
                              disabled: listenInProgress || !hasValidParams,
                              tone: queryInProgress ? "positive" : "primary",
                              text: queryInProgress ? t("action.query-cancel") : t("action.query-execute")
                            })
                          })
                        }), /* @__PURE__ */jsxRuntime.jsx(ui.Box, {
                          flex: 1,
                          marginLeft: 3,
                          children: /* @__PURE__ */jsxRuntime.jsx(ui.Button, {
                            width: "fill",
                            onClick: this.handleListenExecution,
                            type: "button",
                            icon: listenInProgress ? icons.StopIcon : icons.PlayIcon,
                            text: listenInProgress ? t("action.listen-cancel") : t("action.listen-execute"),
                            mode: "ghost",
                            disabled: !hasValidParams,
                            tone: listenInProgress ? "positive" : "default"
                          })
                        })]
                      })
                    })
                  })
                })]
              })]
            })
          }), /* @__PURE__ */jsxRuntime.jsxs(ResultOuterContainer, {
            direction: "column",
            children: [/* @__PURE__ */jsxRuntime.jsx(ResultInnerContainer, {
              flex: 1,
              children: /* @__PURE__ */jsxRuntime.jsx(ResultContainer, {
                flex: 1,
                overflow: "hidden",
                tone: error ? "critical" : "default",
                $isInvalid: Boolean(error),
                children: /* @__PURE__ */jsxRuntime.jsxs(Result, {
                  overflow: "auto",
                  children: [/* @__PURE__ */jsxRuntime.jsx(InputBackgroundContainer, {
                    children: /* @__PURE__ */jsxRuntime.jsx(ui.Box, {
                      marginLeft: 3,
                      children: /* @__PURE__ */jsxRuntime.jsx(StyledLabel, {
                        muted: true,
                        children: t("result.label")
                      })
                    })
                  }), /* @__PURE__ */jsxRuntime.jsxs(ui.Box, {
                    padding: 3,
                    paddingTop: 5,
                    children: [(queryInProgress || listenInProgress && listenMutations.length === 0) && /* @__PURE__ */jsxRuntime.jsx(ui.Box, {
                      marginTop: 3,
                      children: /* @__PURE__ */jsxRuntime.jsx(DelayedSpinner, {})
                    }), error && /* @__PURE__ */jsxRuntime.jsx(QueryErrorDialog, {
                      error
                    }), hasResult && /* @__PURE__ */jsxRuntime.jsx(ResultView, {
                      data: queryResult
                    }), listenInProgress && listenMutations.length > 0 && /* @__PURE__ */jsxRuntime.jsx(ResultView, {
                      data: listenMutations
                    })]
                  })]
                })
              })
            }), /* @__PURE__ */jsxRuntime.jsx(TimingsFooter, {
              children: /* @__PURE__ */jsxRuntime.jsx(TimingsCard, {
                paddingX: 4,
                paddingY: 3,
                sizing: "border",
                children: /* @__PURE__ */jsxRuntime.jsxs(TimingsTextContainer, {
                  align: "center",
                  children: [/* @__PURE__ */jsxRuntime.jsx(ui.Box, {
                    children: /* @__PURE__ */jsxRuntime.jsxs(ui.Text, {
                      muted: true,
                      children: [t("result.execution-time-label"), ":", " ", typeof queryTime === "number" ? "".concat(queryTime, "ms") : t("result.timing-not-applicable")]
                    })
                  }), /* @__PURE__ */jsxRuntime.jsx(ui.Box, {
                    marginLeft: 4,
                    children: /* @__PURE__ */jsxRuntime.jsxs(ui.Text, {
                      muted: true,
                      children: [t("result.end-to-end-time-label"), ":", " ", typeof e2eTime === "number" ? "".concat(e2eTime, "ms") : t("result.timing-not-applicable")]
                    })
                  })]
                })
              })
            })]
          })]
        })
      })]
    });
  }
}
function useDatasets(client) {
  const projectId = client.config().projectId;
  const [datasets, setDatasets] = react.useState();
  react.useEffect(() => {
    const datasets$ = client.observable.datasets.list().subscribe({
      next: result => setDatasets(result.map(ds => ds.name)),
      error: err => setDatasets(err)
    });
    return () => datasets$.unsubscribe();
  }, [client, projectId]);
  return datasets || void 0;
}
function VisionContainer(props) {
  const toast = ui.useToast();
  const loadedDatasets = useDatasets(props.client);
  const {
    t
  } = sanity.useTranslation(index.visionLocaleNamespace);
  if (!loadedDatasets) {
    return /* @__PURE__ */jsxRuntime.jsx(DelayedSpinner, {});
  }
  const datasets = loadedDatasets instanceof Error ?
  // On error, use the clients configured dataset
  [props.client.config().dataset || "production"] :
  // Otherwise use the loaded list, obviously
  loadedDatasets;
  return /* @__PURE__ */jsxRuntime.jsx(VisionGui, {
    ...props,
    datasets,
    toast,
    t
  });
}
class VisionErrorBoundary extends react.Component {
  constructor(props) {
    super(props);
    this.handleRetryRender = () => this.setState(prev => ({
      error: null,
      numRetries: prev.numRetries + 1
    }));
    this.handleRetryWithCacheClear = () => {
      clearLocalStorage();
      this.handleRetryRender();
    };
    this.state = {
      error: null,
      numRetries: 0
    };
  }
  static getDerivedStateFromError(error) {
    return {
      error: error instanceof Error ? error.message : "".concat(error)
    };
  }
  render() {
    if (!this.state.error) {
      return this.props.children;
    }
    const message = this.state.error;
    const withCacheClear = this.state.numRetries > 0;
    return /* @__PURE__ */jsxRuntime.jsx(ui.Card, {
      height: "fill",
      overflow: "auto",
      paddingY: [4, 5, 6, 7],
      paddingX: 4,
      sizing: "border",
      tone: "critical",
      children: /* @__PURE__ */jsxRuntime.jsx(ui.Container, {
        width: 3,
        children: /* @__PURE__ */jsxRuntime.jsxs(ui.Stack, {
          space: 4,
          children: [/* @__PURE__ */jsxRuntime.jsx("div", {
            children: /* @__PURE__ */jsxRuntime.jsx(ui.Button, {
              onClick: withCacheClear ? this.handleRetryWithCacheClear : this.handleRetryRender,
              text: withCacheClear ? "Clear cache and retry" : "Retry",
              tone: "default"
            })
          }), /* @__PURE__ */jsxRuntime.jsx(ui.Heading, {
            children: "An error occured"
          }), /* @__PURE__ */jsxRuntime.jsx(ui.Card, {
            border: true,
            radius: 2,
            overflow: "auto",
            padding: 4,
            tone: "inherit",
            children: /* @__PURE__ */jsxRuntime.jsx(ui.Stack, {
              space: 4,
              children: message && /* @__PURE__ */jsxRuntime.jsx(ui.Code, {
                size: 1,
                children: /* @__PURE__ */jsxRuntime.jsxs("strong", {
                  children: ["Error: ", message]
                })
              })
            })
          })]
        })
      })
    });
  }
}
function SanityVision(props) {
  const client = sanity.useClient({
    apiVersion: "1"
  });
  const config = {
    defaultApiVersion: DEFAULT_API_VERSION,
    ...props.tool.options
  };
  return /* @__PURE__ */jsxRuntime.jsx(VisionErrorBoundary, {
    children: /* @__PURE__ */jsxRuntime.jsx(VisionContainer, {
      client,
      config
    })
  });
}
exports.default = SanityVision;
