// @bjoerge/mutiny/dist/_path.cjs
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var stringify = require('./_chunks/stringify-Fw2FifQ5.cjs');
var parse = require('./_chunks/parse-8ISYfcYA.cjs');
function getAtPath(path, value) {
  if (path.length === 0) {
    return value;
  }
  let current = value;
  for (const head of path) {
    if (stringify.isArrayElement(head)) {
      if (!Array.isArray(current)) {
        return void 0;
      }
      if (stringify.isKeyedElement(head)) {
        current = current.find(item => item._key === head._key);
        continue;
      }
      current = current[head];
      continue;
    }
    current = current[head];
  }
  return current;
}
function normalize(path) {
  return typeof path === "string" ? parse.parse(path) : path;
}
exports.isArrayElement = stringify.isArrayElement;
exports.isElementEqual = stringify.isElementEqual;
exports.isEqual = stringify.isEqual;
exports.isIndexElement = stringify.isIndexElement;
exports.isKeyElement = stringify.isKeyElement;
exports.isKeyedElement = stringify.isKeyedElement;
exports.isPropertyElement = stringify.isPropertyElement;
exports.startsWith = stringify.startsWith;
exports.stringify = stringify.stringify;
exports.parse = parse.parse;
exports.getAtPath = getAtPath;
exports.normalize = normalize;
